<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>Variant Calling Workflow</title>
        
            <script async defer data-domain="training.galaxyproject.org" src="https://plausible.galaxyproject.eu/js/plausible.js"></script>

        
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap-toc.min.css">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <script src="https://kit.fontawesome.com/67b3f98409.js" crossorigin="anonymous"></script>
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />
        <link rel="alternate" type="application/atom+xml" href="/training-material/feed.xml" />

        
        
        
        
        
        <meta name="description" content="Genetic differences (variants) between healthy and diseas..." />
        <meta property="og:title" content="Galaxy Training: Variant Calling Workflow" />
        <meta property="og:description" content="Genetic differences (variants) between healthy and diseas..." />
        <meta property="og:image" content="/training-material/assets/images/GTNLogo1000.png" />
    </head>
    <body data-spy="scroll" data-target="#toc">
        <header>
    <nav class="navbar navbar-expand-md navbar-dark" aria-label="Site Navigation">
        <div class="container">
            <a class="navbar-brand" href="/training-material/">
                <img src="/training-material/assets/images/GTN-60px.png" height="30" alt="Galaxy Training Network logo">
                Galaxy Training!
            </a>

            <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#top-navbar" aria-controls="top-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        
                        <a class="nav-link" href="/training-material/topics/variant-analysis" title="Go back to list of tutorials">
                            <i class="far fa-folder" aria-hidden="true"></i> Variant Analysis
                        </a>
                        
                    </li>

                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Languages">
                            <i class="fas fa-language" aria-hidden="true"></i><span class="visually-hidden">language</span> Languages
                        </a>
                        <div class="dropdown-menu dropdown-menu-right">
                            
                              <strong class="dropdown-header">Automatic Translations</strong>
            
            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=fr&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/variant-analysis/tutorials/variant-calling/tutorial.html&edit-text=&act=url">
                            Fran√ßais
                            </a>
                            
            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ja&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/variant-analysis/tutorials/variant-calling/tutorial.html&edit-text=&act=url">
                            Êó•Êú¨Ë™û
                            </a>
                            
            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=es&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/variant-analysis/tutorials/variant-calling/tutorial.html&edit-text=&act=url">
                            Espa√±ol
                            </a>
                            
            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=pt&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/variant-analysis/tutorials/variant-calling/tutorial.html&edit-text=&act=url">
                            Portugu√™s
                            </a>
                            
            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ar&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/variant-analysis/tutorials/variant-calling/tutorial.html&edit-text=&act=url">
                            ÿßŸÑÿπÿ±ÿ®Ÿäÿ©
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/variant-analysis/tutorials/variant-calling/tutorial.html&edit-text=&act=url" title="">
                            And more!
                            </a>
                        </div>
                    </li>

                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu dropdown-menu-right">
        <!-- disable Tess for now
        <form method="get" action="https://tess.elixir-europe.org/materials">
            <input type="text" id="search" name="q" value="" style="margin-left: 0.5em;/*! border-radius: 0px; */">
            <input type="hidden" value="Galaxy Training" name="content_provider">
            <input type="submit" value="Search on TeSS" style="width: 92%;border-radius: 0px;margin: 0.5em;background: #f47d20;border: 0px;padding: 0.25em;" class="">
        </form>
        -->

        <a class="dropdown-item" href="/training-material/faq" title="Check our FAQs">
           <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> FAQs
        </a>
        
        <a class="dropdown-item" href="/training-material/topics/variant-analysis/faqs/" title="Check our FAQs for the Variant Analysis topic">
           <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Topic FAQs
        </a>
        
        <a class="dropdown-item" href="https://help.galaxyproject.org/" title="Discuss on Galaxy Help">
            <i class="far fa-comments" aria-hidden="true"></i><span class="visually-hidden">feedback</span> Galaxy Help Forum
        </a>
        <a class="dropdown-item" href="https://gitter.im/Galaxy-Training-Network/Lobby" title="Discuss on gitter">
           <i class="fab fa-gitter" aria-hidden="true"></i><span class="visually-hidden">gitter</span> Discuss on Gitter
        </a>
    </div>
</li>


                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Extras">
        <i class="far fa-star" aria-hidden="true"></i><span class="visually-hidden">galaxy-star</span> Extras
    </a>
    <div class="dropdown-menu dropdown-menu-right">

        
        <a class="dropdown-item" href="https://github.com/galaxyproject/training-material/edit/main/topics/variant-analysis/tutorials/variant-calling/tutorial.md" title="Edit on GitHub">
          <i class="fab fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> Edit on GitHub
        </a>

        <a class="dropdown-item" href="/training-material/stats.html" title="Show view statistics about this repository">
            <i class="fas fa-chart-bar" aria-hidden="true"></i><span class="visually-hidden">galaxy-barchart</span> GTN statistics
        </a>

        <a class="dropdown-item" href="https://plausible.galaxyproject.eu/training.galaxyproject.org?period=12mo&page=/training-material/topics/variant-analysis/tutorials/variant-calling/tutorial.html" title="Show view statistics of this page">
            <i class="fas fa-chart-bar" aria-hidden="true"></i><span class="visually-hidden">galaxy-barchart</span> Page Metrics
        </a>

        
            <a class="dropdown-item" href="/training-material/feedback.html" title="Show feedback statistics about this repository">
                <i class="fas fa-chart-bar" aria-hidden="true"></i><span class="visually-hidden">galaxy-barchart</span> GTN feedback
            </a>
        

        <div class="dropdown-item">
            <div>
                <i class="fas fa-palette" aria-hidden="true"></i><span class="visually-hidden">gtn-theme</span> Theme
            </div>

            <div id="theme-selector" data-toggle="buttons">
                <label data-value="default" class="btn btn-secondary">
                    <input type="radio" name="options" id="default" autocomplete="off"> Default
                </label>
                <label data-value="night" class="btn btn-secondary">
                    <input type="radio" name="options" id="night" autocomplete="off"> Night
                </label>
                <label data-value="midnight" class="btn btn-secondary">
                    <input type="radio" name="options" id="midnight" autocomplete="off"> Midnight
                </label>
                <label data-value="rainbow" class="btn btn-secondary">
                    <input type="radio" name="options" id="rainbow" autocomplete="off"> Rainbow
                </label>
                <label data-value="progress" class="btn btn-secondary">
                    <input type="radio" name="options" id="progress" autocomplete="off">üè≥Ô∏è‚Äçüåà
                </label>
                <label data-value="halloween" class="btn btn-secondary">
                    <input type="radio" name="options" id="halloween" autocomplete="off"> üéÉ
                </label>
                <label data-value="straya" class="btn btn-secondary">
                    <input type="radio" name="options" id="downunder" autocomplete="off"> üá¶üá∫
                </label>
            </div>

        </div>

        <div class="dropdown-item">
            <div>
                <i class="fas fa-history" aria-hidden="true"></i><span class="visually-hidden">galaxy-rulebuilder-history</span> Previous Versions
            </div>

            
            <div id="archive-selector">
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/2021-09-01/topics/variant-analysis/tutorials/variant-calling/tutorial.html" title="Version 2021-09-01">2021-09-01</a>
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/2021-08-01/topics/variant-analysis/tutorials/variant-calling/tutorial.html" title="Version 2021-08-01">2021-08-01</a>
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/2021-07-01/topics/variant-analysis/tutorials/variant-calling/tutorial.html" title="Version 2021-07-01">2021-07-01</a>
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/" title="Older Versions">Older Versions</a>
            </div>

        </div>

    </div>
</li>


                    <!-- Search bar-->
                    <li class="nav-item">
                      <div id="navbarSupportedContent" role="search">
                        <!-- Search form -->
                        <form class="form-inline mr-auto" method="GET" action="/training-material/search">
                          <i class="fas fa-search nav-link" aria-hidden="true"></i>
                          <div class="md-form mb-2">
                            <input name="query" class="form-control nicer" type="text" placeholder="Search Tutorials" aria-label="Search">
                          </div>
                        </form>
                      </div>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
</header>

        <div class="container main-content" role="main">
        











<!-- Gitter -->




<script>
  ((window.gitter = {}).chat = {}).options = {
  room: 'Galaxy-Training-Network/Lobby'
  };
</script>
<script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>

<script type="application/ld+json">
    


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "citation": {
    "@type": "CreativeWork",
    "name": "Community-Driven Data Analysis Training for Biology",
    "url": "https://doi.org/10.1016/j.cels.2018.05.012"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "dateModified": "1970-01-01 00:33:42 +0000",
  "discussionUrl": "https://gitter.im/Galaxy-Training-Network/Lobby",
  "headline": "Variant Calling Workflow",
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "isFamilyFriendly": true,
  "license": "https://spdx.org/licenses/CC-BY-4.0.html",
  "producer": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "provider": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "isPartOf": {
    "@type": "CreativeWork",
    "name": "Variant Analysis",
    "description": "Genetic differences (variants) between healthy and diseased tissue,\nbetween individuals of a population, or between strains of an organism\ncan provide mechanistic insight into disease processes and the natural\nfunction of affected genes.\n\nThe tutorials in this section show how to detect evidence for\ngenetic variants in next-generation sequencing data, a process termed\nvariant calling.\n\nOf equal importance, they also demonstrate how you can interpret, for\na range of different organisms, the resulting sets of variants by\npredicting their molecular effects on genes and proteins, by\nannotating previously observed variants with published knowledge, and\nby trying to link phenotypes of the sequenced samples to their variant\ngenotypes.\n",
    "url": "https://training.galaxyproject.org//training-material/topics/variant-analysis/"
  },
  "courseCode": "variant-analysis / variant-calling / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for 'Variant Calling Workflow' tutorial",
  "url": "https://training.galaxyproject.org//training-material/topics/variant-analysis/tutorials/variant-calling/tutorial.html",
  "timeRequired": "PT50 MIN",
  "description": "The questions this  addresses are:\n - How do I find sequence variants between my sample and a reference genome?\n\n\\nThe objectives are:\n - Understand the steps involved in variant calling.\n - Describe the types of data formats encountered during variant calling.\n - Use command line tools to perform variant calling.\n\n",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "coursePrerequisites": [
    {
      "@type": "CreativeWork",
      "url": "https://training.galaxyproject.org//training-material/topics/introduction/",
      "name": "Introduction to Galaxy Analyses",
      "description": "Introduction to Galaxy Analyses",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/quality-control/slides.html",
      "name": "Quality Control",
      "description": "Slides for 'Quality Control' tutorial",
      "learningResourceType": "slides",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html",
      "name": "Quality Control",
      "description": "Hands-on for 'Quality Control' tutorial",
      "learningResourceType": "hands-on tutorial",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/quality-control/slides.html",
      "name": "Quality Control",
      "description": "Slides for 'Quality Control' tutorial",
      "learningResourceType": "slides",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/mapping/slides.html",
      "name": "Mapping",
      "description": "Slides for 'Mapping' tutorial",
      "learningResourceType": "slides",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/mapping/tutorial.html",
      "name": "Mapping",
      "description": "Hands-on for 'Mapping' tutorial",
      "learningResourceType": "hands-on tutorial",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/mapping/slides.html",
      "name": "Mapping",
      "description": "Slides for 'Mapping' tutorial",
      "learningResourceType": "slides",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    }
  ],
  "hasPart": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Sofoklis Keisaris"
    },
    {
      "@type": "Person",
      "name": "Fotis E. Psomopoulos"
    },
    {
      "@type": "Person",
      "name": "Saskia Hiltemann"
    },
    {
      "@type": "Person",
      "name": "Helena Rasche"
    },
    {
      "@type": "Person",
      "name": "The Carpentries"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Sofoklis Keisaris"
    },
    {
      "@type": "Person",
      "name": "Fotis E. Psomopoulos"
    },
    {
      "@type": "Person",
      "name": "Saskia Hiltemann"
    },
    {
      "@type": "Person",
      "name": "Helena Rasche"
    },
    {
      "@type": "Person",
      "name": "The Carpentries"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Variant Analysis",
      "description": "Genetic differences (variants) between healthy and diseased tissue,\nbetween individuals of a population, or between strains of an organism\ncan provide mechanistic insight into disease processes and the natural\nfunction of affected genes.\n\nThe tutorials in this section show how to detect evidence for\ngenetic variants in next-generation sequencing data, a process termed\nvariant calling.\n\nOf equal importance, they also demonstrate how you can interpret, for\na range of different organisms, the resulting sets of variants by\npredicting their molecular effects on genes and proteins, by\nannotating previously observed variants with published knowledge, and\nby trying to link phenotypes of the sequenced samples to their variant\ngenotypes.\n",
      "url": "https://training.galaxyproject.org//training-material/topics/variant-analysis/"
    },
    {
      "@type": "DefinedTerm",
      "@id": "http://edamontology.org/topic_0199",
      "inDefinedTermSet": "http://edamontology.org",
      "termCode": "topic_0199",
      "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Ftopic_0199"
    }
  ]
}
</script>

<section class="tutorial topic-variant-analysis">
    <h1 data-toc-skip>Variant Calling Workflow</h1>
    


    <div class="contributors-line">Authors: 
<a href="/training-material/hall-of-fame/Sofokli5/" class="contributor-badge contributor-Sofokli5"><img src="https://avatars.githubusercontent.com/Sofokli5?s=27" alt="Avatar">Sofoklis Keisaris</a>


<a href="/training-material/hall-of-fame/fpsom/" class="contributor-badge contributor-fpsom"><img src="https://avatars.githubusercontent.com/fpsom?s=27" alt="Avatar">Fotis E. Psomopoulos</a>


<a href="/training-material/hall-of-fame/shiltemann/" class="contributor-badge contributor-shiltemann"><img src="https://avatars.githubusercontent.com/shiltemann?s=27" alt="Avatar">Saskia Hiltemann</a>


<a href="/training-material/hall-of-fame/hexylena/" class="contributor-badge contributor-hexylena"><img src="https://avatars.githubusercontent.com/hexylena?s=27" alt="Avatar">Helena Rasche</a>


<a href="/training-material/hall-of-fame/carpentries/" class="contributor-badge contributor-carpentries"><img src="https://avatars.githubusercontent.com/carpentries?s=27" alt="Avatar">The Carpentries</a>


<a href="https://training.galaxyproject.org/training-material/topics/contributing/" class="contributor-badge contributor-newcontributors"><i class="fas fa-users" aria-hidden="true"></i><span class="visually-hidden">hall-of-fame</span>Add Contributions!</a>

</div>


    <blockquote class="overview">
        <h3>Overview</h3>
        
        <img alt="Creative Commons License" class="float-right" style="border-width:0; display: inline-block; margin:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" aria-hidden="true" />
        
        <strong><i class="far fa-question-circle" aria-hidden="true"></i> Questions:</strong>
        <ul>
        
        <li><p>How do I find sequence variants between my sample and a reference genome?</p>
</li>
        
        </ul>

        <strong><i class="fas fa-bullseye" aria-hidden="true"></i> Objectives: </strong>
        <ul>
        
        <li><p>Understand the steps involved in variant calling.</p>
</li>
        
        <li><p>Describe the types of data formats encountered during variant calling.</p>
</li>
        
        <li><p>Use command line tools to perform variant calling.</p>
</li>
        
        </ul>

        
        <strong><i class="fas fa-check-circle" aria-hidden="true"></i> Requirements:</strong>
        <ul>
        
    <li>
    
        
        
        <a href="/training-material/topics/introduction">Introduction to Galaxy Analyses</a>
        
    
    </li>

    <li>
    
        
        
        <a href="/training-material/topics/sequence-analysis">Sequence analysis</a>
        
            <ul>
                
                    
                        
                    
                        
                            
                            <li> Quality Control:
                            
                                <a href="/training-material/topics/sequence-analysis/tutorials/quality-control/slides.html"><i class="fab fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> slides</a>
                                
                            
                            
                                
                                    - <a href="/training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html"><i class="fas fa-laptop" aria-hidden="true"></i><span class="visually-hidden">tutorial</span> hands-on</a>
                                
                            
                            </li>
                        
                    
                        
                    
                
                    
                        
                            
                            <li> Mapping:
                            
                                <a href="/training-material/topics/sequence-analysis/tutorials/mapping/slides.html"><i class="fab fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> slides</a>
                                
                            
                            
                                
                                    - <a href="/training-material/topics/sequence-analysis/tutorials/mapping/tutorial.html"><i class="fas fa-laptop" aria-hidden="true"></i><span class="visually-hidden">tutorial</span> hands-on</a>
                                
                            
                            </li>
                        
                    
                        
                    
                        
                    
                
            </ul>
        
    
    </li>

        
        </ul>
        

        
        <div><strong><i class="fas fa-hourglass-half" aria-hidden="true"></i> Time estimation:</strong> 50 min</div>
        

        

        
        

        
        <div id="supporting-materials"><strong><i class="fa fa-external-link" aria-hidden="true"></i> Supporting Materials:</strong></div>
        <ul class="supporting_material">
            
                <li class="btn btn-default supporting_material"><a href="/training-material/topics/variant-analysis/slides/introduction.html" title="Topic Overview slides">
                    <i class="fab fa-slideshare" aria-hidden="true"></i> Topic Overview slides
                </a></li>
            

            

            

            

            

            
            
            

            <!-- Check the GTN Video Library for recordings of this tutorial or associated slides -->
            













  



            
                <li class="btn btn-default supporting_material">


    <a href="#" class="btn btn-default dropdown-toggle topic-icon" data-toggle="dropdown" aria-expanded="false" title="Where to run the tutorial">
        <i class="fas fa-globe" aria-hidden="true"></i><span class="visually-hidden">instances</span>&nbsp;Available on these Galaxies 
    </a>
    <ul class="dropdown-menu">
    
        <li>
            <a class="dropdown-item" href="https://github.com/galaxyproject/training-material/tree/main/topics/variant-analysis/docker" title="Docker image for this tutorial">
                <i class="fab fa-docker" aria-hidden="true"></i><span class="visually-hidden">docker_image</span> Docker image
            </a>
        </li>
    
    
    </ul>


</li>
            
        </ul>
        

        <div><strong><i class="far fa-calendar" aria-hidden="true"></i> Last modification:</strong> May 18, 2022 </div>
        <div><strong><i class="fas fa-balance-scale" aria-hidden="true"></i> License:</strong>
            
            <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Tutorial Content is licensed under Creative Commons Attribution 4.0 International License</a>
            
            <a rel="license" href="https://github.com/galaxyproject/training-material/blob/main/LICENSE.md">The GTN Framework is licensed under MIT</a>
        </div>
    </blockquote>

    <div class="container">
        <div class="row">
            <!-- sidebar, which will move to the top on a small screen -->
            <div class="col-sm-2">
                <nav id="toc" data-toggle="toc" class="sticky-top" aria-label="Table of Contents"></nav>
            </div>
            <div class="col-sm-10">
                 

                <h1 class="no_toc" id="introduction">Introduction</h1>

<blockquote class="comment">
  <h3 id="comment-comment"><i class="far fa-comment-dots" aria-hidden="true"></i><span class="visually-hidden">comment</span> Comment</h3>

  <p>This tutorial is significantly based on <a href="https://carpentries.org">the Carpentries</a> <a href="https://datacarpentry.org/wrangling-genomics/">‚ÄúData Wrangling and Processing for Genomics‚Äù</a> lesson</p>

</blockquote>

<p>In this tutorial we are working with files from a long-term evolution study of an <em>E. coli</em> population (designated Ara-3). We will perform variant calling to see how the population changed over time. We care how this population changed relative to the original population, <em>E. coli</em> strain REL606. Therefore, we will align each of our samples to the <em>E. coli</em> REL606 reference genome, and see what differences exist in our <span class="notranslate">reads</span> versus the genome.</p>

<p>The whole proccess is designed to be run in <code class="language-plaintext highlighter-rouge">RStudio</code> ,as an <code class="language-plaintext highlighter-rouge">Active Interactive Tool</code>, directly in <code class="language-plaintext highlighter-rouge"><span class="notranslate">Galaxy</span></code>.</p>

<h1 id="hands_on-installing-software"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Installing software</h1>

<p>First, we need to get our hands on RStudio.</p>

<blockquote class="tip">
  <h3 id="tip-tip-launch-rstudio-in-galaxy"><i class="far fa-lightbulb" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Launch RStudio in <span class="notranslate">Galaxy</span></h3>
  <p>Currently RStudio in <span class="notranslate">Galaxy</span> is only available on <a href="https://usegalaxy.eu">Use<span class="notranslate">Galaxy</span>.eu</a> and <a href="https://usegalaxy.org">Use<span class="notranslate">Galaxy</span>.org</a></p>

  <ol>
    <li>Open the Rstudio tool <i class="fas fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> by clicking <a href="https://usegalaxy.eu/?tool_id=interactive_tool_rstudio">here</a></li>
    <li>Click Execute</li>
    <li>The tool will start running and will stay running permanently</li>
    <li>Click on the ‚ÄúUser‚Äù menu at the top and go to ‚ÄúActive InteractiveTools‚Äù and locate the RStudio instance you started.</li>
  </ol>
</blockquote>
<blockquote class="tip">

  <h3 id="tip-tip-launch-rstudio-cloud-if-not-available-on-galaxy"><i class="far fa-lightbulb" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Launch RStudio Cloud if not available on <span class="notranslate">Galaxy</span></h3>

  <p>If RStudio is not available on the <span class="notranslate">Galaxy</span> instance:</p>
  <ol>
    <li>Register for <a href="https://client.login.rstudio.cloud/oauth/login?show_auth=0&amp;show_login=1&amp;show_setup=1">RStudio Cloud</a>, or login if you already have an account</li>
    <li>Create a new project</li>
  </ol>
</blockquote>

<p>The R Console and other interactive tools like RStudio are great for prototyping code and exploring data, but sooner or later we will want to use our program in a pipeline or run it in a shell script to process thousands of data files. This is one of those cases and, in order to do that, we will use the terminal provided by the RStudio itself. 
We go to ‚ÄúTools‚Äù and pick the ‚ÄúShell‚Ä¶‚Äù option and we are good to go. Our workspace is the left, terminal window that just opened.</p>

<p>Fortunately, <a href="https://docs.conda.io/en/latest/miniconda.html">miniconda</a> is already installed. Miniconda is a package manager that simplifies the installation processes. We can and will use it to install every essential package for our tutorial. However, it is of critical importance that we do that in an new environment within our existing base and install our packages in said environment.</p>

<blockquote class="code-in">
  <h2 id="code-in-environment-and-packages"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> Environment and Packages</h2>
  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>conda create <span class="nt">-n</span> name_of_your_env bwa samtools bcftools
<span class="nv">$ </span>conda activate name_of_your_env
</code></pre></div>  </div>
</blockquote>

<table>
  <thead>
    <tr>
      <th>Software</th>
      <th>Version</th>
      <th>Manual</th>
      <th>Available for</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://bio-bwa.sourceforge.net/">BWA</a></td>
      <td>0.7.17</td>
      <td><a href="http://bio-bwa.sourceforge.net/bwa.shtml">Link</a></td>
      <td>Linux, MacOS</td>
      <td><span class="notranslate">Mapping</span> <span class="notranslate">DNA</span> sequences against reference genome.</td>
    </tr>
    <tr>
      <td><a href="http://samtools.sourceforge.net/">SAMtools</a></td>
      <td>1.15.1</td>
      <td><a href="http://www.htslib.org/doc/samtools.html">Link</a></td>
      <td>Linux, MacOS</td>
      <td>Utilities for manipulating alignments in the SAM format.</td>
    </tr>
    <tr>
      <td><a href="https://samtools.github.io/bcftools/">BCFtools</a></td>
      <td>1.15.1</td>
      <td><a href="https://samtools.github.io/bcftools/bcftools.html">Link</a></td>
      <td>Linux, MacOS</td>
      <td>Utilities for variant calling and manipulating VCFs and BCFs.</td>
    </tr>
    <tr>
      <td><a href="http://software.broadinstitute.org/software/igv/home">IGV</a></td>
      <td><a href="https://software.broadinstitute.org/software/igv/download">Link</a></td>
      <td><a href="https://software.broadinstitute.org/software/igv/UserGuide">Link</a></td>
      <td>Linux, MacOS, Windows</td>
      <td>Visualization and interactive exploration of large genomics datasets.</td>
    </tr>
  </tbody>
</table>

<h1 class="hands_on" id="alignment-to-a-reference-genome">Alignment to a reference genome</h1>

<p><img src="../../images/variant_calling_workflow_align.png" alt="workflow_align. " loading="lazy" /></p>

<p>We perform read alignment or mappinŒ≥ to determine where in the genome our <span class="notranslate">reads</span> originated from. There are a number of tools to
choose from and, while there is no gold standard, there are some tools that are better suited for particular NGS analyses. We will be
using the <a href="http://bio-bwa.sourceforge.net/">Burrows Wheeler Aligner (BWA)</a>, which is a software package for <span class="notranslate">mapping</span> low-divergent
sequences against a large reference genome.</p>

<p>The alignment process consists of two steps:</p>

<ol>
  <li>Indexing the reference genome</li>
  <li>Aligning the <span class="notranslate">reads</span> to the reference genome</li>
</ol>

<h1 id="setting-up">Setting up</h1>

<blockquote class="notranslate hands_on">
  <h2 id="hands_on-download-genome"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Download Genome</h2>

  <p>First we download the reference genome for <em>E. coli</em> REL606. Although we could copy or move the file with <code class="language-plaintext highlighter-rouge">cp</code> or <code class="language-plaintext highlighter-rouge">mv</code>, most genomics workflows begin with a download step, so we will practice that here.</p>

  <blockquote>
    <h3 id="code-in-input-bash"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> Input: Bash</h3>
    <div class="language-bash code-in highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">mkdir</span> ~/<span class="notranslate">galaxy</span>_tutorial
<span class="nv">$ </span><span class="nb">cd</span> ~/<span class="notranslate">galaxy</span>_tutorial
<span class="nv">$ </span><span class="nb">mkdir</span> <span class="nt">-p</span> data/ref_genome
<span class="nv">$ </span>curl <span class="nt">-L</span> <span class="nt">-o</span> data/ref_genome/ecoli_rel606.fasta.gz ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/017/985/GCA_000017985.1_ASM1798v1/GCA_000017985.1_ASM1798v1_genomic.fna.gz
<span class="nv">$ </span><span class="nb">gunzip </span>data/ref_genome/ecoli_rel606.fasta.gz
</code></pre></div>    </div>
  </blockquote>

  <blockquote>
    <h3 id="question-question"><i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>We saved this file as <code class="language-plaintext highlighter-rouge">data/ref_genome/ecoli_rel606.fasta.gz</code> and then decompressed it. 
What is the real name of the genome?</p>

    <blockquote>
      <h3 id="solution-solution"><i class="far fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

      <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">head </span>data/ref_genome/ecoli_rel606.fasta
</code></pre></div>      </div>

      <p>The name of the sequence follows the <code class="language-plaintext highlighter-rouge">&gt;</code> character. The name is <code class="language-plaintext highlighter-rouge">CP000819.1 Escherichia coli B str. REL606, complete genome</code>.
Keep this chromosome name (<code class="language-plaintext highlighter-rouge">CP000819.1</code>) in mind, as we will use it later in the lesson.</p>
    </blockquote>

  </blockquote>

  <p class="question">We will also download a set of trimmed FASTQ files to work with that will enable us to run our variant calling <span class="notranslate">workflow</span> quite quickly.</p>

  <blockquote>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>curl <span class="nt">-L</span> <span class="nt">-o</span> sub.tar.gz https://ndownloader.figshare.com/files/14418248
<span class="nv">$ </span><span class="nb">tar </span>xvf sub.tar.gz
<span class="nv">$ </span><span class="nb">mv </span>sub/ ~/<span class="notranslate">galaxy</span>_tutorial/data/trimmed_fastq_small
</code></pre></div>    </div>

    <p>You will also need to create directories for the results that will be generated as part of this <span class="notranslate">workflow</span>. We can do this in a single line of code, because <code class="language-plaintext highlighter-rouge">mkdir</code> can accept multiple new directory names as input.</p>
  </blockquote>

  <blockquote>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">mkdir</span> <span class="nt">-p</span> results/sam results/bam results/bcf results/vcf
</code></pre></div>    </div>
  </blockquote>

</blockquote>

<blockquote class="notranslate hands_on">
  <h2 id="hands_on-index-the-reference-genome"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Index the reference genome</h2>
  <p>Our first step is to index the reference genome for use by BWA. Indexing allows the aligner to quickly find potential alignment sites for query sequences in a genome, which saves time during alignment. Indexing the reference only has to be run once. The only reason you would want to create a new index is if you are working with a different reference genome or you are using a different tool for alignment.</p>

  <blockquote>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>bwa index data/ref_genome/ecoli_rel606.fasta
</code></pre></div>    </div>
  </blockquote>

  <h3 id="code-out-output"><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span> Output</h3>
  <p>While the index is created, you will see output that looks something like this:</p>

  <blockquote class="icon">
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[bwa_index] Pack FASTA... 0.04 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 1.05 seconds elapse.
[bwa_index] Update BWT... 0.03 sec
[bwa_index] Pack forward-only FASTA... 0.02 sec
[bwa_index] Construct SA from BWT and Occ... 0.57 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index data/ref_genome/ecoli_rel606.fasta
[main] Real time: 1.765 sec; CPU: 1.715 sec
</code></pre></div>    </div>
  </blockquote>

</blockquote>

<blockquote class="icon">
  <h2 id="hands_on-align-reads-to-reference-genome"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Align <span class="notranslate">reads</span> to reference genome</h2>

  <p>The alignment process consists of choosing an appropriate reference genome to map our <span class="notranslate">reads</span> against and then deciding on an 
aligner. We will use the BWA-MEM algorithm, which is the latest and is generally recommended for high-quality queries as it 
is faster and more accurate.</p>

  <p>An example of what a <code class="language-plaintext highlighter-rouge">bwa</code> command looks like is below. This command will not run, as we do not have the files <code class="language-plaintext highlighter-rouge">ref_genome.fa</code>, <code class="language-plaintext highlighter-rouge">input_file_R1.fastq</code>, or <code class="language-plaintext highlighter-rouge">input_file_R2.fastq</code>.</p>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>bwa mem ref_genome.fasta input_file_R1.fastq input_file_R2.fastq <span class="o">&gt;</span> output.sam
</code></pre></div>  </div>

  <p>Have a look at the <a href="http://bio-bwa.sourceforge.net/bwa.shtml">bwa options page</a>. While we are running bwa with the default 
parameters here, your use case might require a change of parameters. <em>NOTE: Always read the manual page for any tool before using 
and make sure the options you use are appropriate for your data.</em></p>

  <p>We are going to start by aligning the <span class="notranslate">reads</span> from just one of the 
samples in our dataset (<code class="language-plaintext highlighter-rouge">SRR2584866</code>). Later, we will be 
iterating this whole process on all of our sample files.</p>
  <blockquote class="icon">
    <h3 id="code-in-align-with-bwa"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> Align with bwa</h3>
    <div class="language-bash icon highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>bwa mem data/ref_genome/ecoli_rel606.fasta data/trimmed_fastq_small/SRR2584866_1.trim.sub.fastq data/trimmed_fastq_small/SRR2584866_2.trim.sub.fastq <span class="o">&gt;</span> results/sam/SRR2584866.aligned.sam
</code></pre></div>    </div>

    <h3 id="-icon-code-out-output">{: .icon code-out} Output:</h3>

    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">[</span>M::bwa_idx_load_from_disk] <span class="nb">read </span>0 ALT contigs
<span class="o">[</span>M::process] <span class="nb">read </span>77446 sequences <span class="o">(</span>10000033 bp<span class="o">)</span>...
<span class="o">[</span>M::process] <span class="nb">read </span>77296 sequences <span class="o">(</span>10000182 bp<span class="o">)</span>...
<span class="o">[</span>M::mem_pestat] <span class="c"># candidate unique pairs for (FF, FR, RF, RR): (48, 36728, 21, 61)</span>
<span class="o">[</span>M::mem_pestat] analyzing insert size distribution <span class="k">for </span>orientation FF...
<span class="o">[</span>M::mem_pestat] <span class="o">(</span>25, 50, 75<span class="o">)</span> percentile: <span class="o">(</span>420, 660, 1774<span class="o">)</span>
<span class="o">[</span>M::mem_pestat] low and high boundaries <span class="k">for </span>computing mean and std.dev: <span class="o">(</span>1, 4482<span class="o">)</span>
<span class="o">[</span>M::mem_pestat] mean and std.dev: <span class="o">(</span>784.68, 700.87<span class="o">)</span>
<span class="o">[</span>M::mem_pestat] low and high boundaries <span class="k">for </span>proper pairs: <span class="o">(</span>1, 5836<span class="o">)</span>
<span class="o">[</span>M::mem_pestat] analyzing insert size distribution <span class="k">for </span>orientation FR...
</code></pre></div>    </div>
  </blockquote>
</blockquote>
<blockquote class="notranslate hands_on">

  <h2 id="hands_on-sambam-format"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> SAM/BAM format</h2>
  <p>The <a href="https://genome.sph.umich.edu/wiki/SAM">SAM file</a>,
is a tab-delimited text file that contains information for each individual read and its alignment to the genome. While we do not 
have time to go into detail about the features of the SAM format, the paper by 
<a href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Heng Li et al.</a> provides a lot more detail on the specification.</p>

  <p><strong>The compressed binary version of SAM is called a BAM file.</strong> We use this version to reduce size and to allow for <em>indexing</em>, which enables efficient random access of the data contained within the file.</p>

  <p>The file begins with a <strong>header</strong>, which is optional. The header is used to describe the source of data, reference sequence, method of
alignment, etc., this will change depending on the aligner being used. Following the header is the <strong>alignment section</strong>. Each line
that follows corresponds to alignment information for a single read. Each alignment line has <strong>11 mandatory fields</strong> for essential
<span class="notranslate">mapping</span> information and a variable number of other fields for aligner specific information. An example entry from a SAM file is 
displayed below with the different fields highlighted.</p>

  <p><img src="../../images/sam_bam.png" alt="sam_bam1. " loading="lazy" /></p>

  <p><img src="../../images/sam_bam3.png" alt="sam_bam2. " loading="lazy" /></p>

  <blockquote class="icon">
    <h3 id="code-in-sam-to-bam"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> SAM to BAM</h3>
    <p>We will convert the SAM file to BAM format using the <code class="language-plaintext highlighter-rouge">samtools</code> program with the <code class="language-plaintext highlighter-rouge">view</code> command and tell this command that the input is in SAM format (<code class="language-plaintext highlighter-rouge">-S</code>) and to output BAM format (<code class="language-plaintext highlighter-rouge">-b</code>):</p>

    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>samtools view <span class="nt">-S</span> <span class="nt">-b</span> results/sam/SRR2584866.aligned.sam <span class="o">&gt;</span> results/bam/SRR2584866.aligned.bam
</code></pre></div>    </div>
  </blockquote>
  <blockquote class="icon">
    <h3 id="code-out-output-1"><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span> Output</h3>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">[</span>samopen] SAM header is present: 1 sequences.
</code></pre></div>    </div>
  </blockquote>

</blockquote>

<blockquote>
  <h2 id="hands_on-sort-bam-file-by-coordinates"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Sort BAM file by coordinates</h2>

  <p>Next we sort the BAM file using the <code class="language-plaintext highlighter-rouge">sort</code> command from <code class="language-plaintext highlighter-rouge">samtools</code>. <code class="language-plaintext highlighter-rouge">-o</code> tells the command where to write the output.</p>
  <blockquote class="icon">
    <h3 id="code-in-sort-command"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> <code class="language-plaintext highlighter-rouge">sort</code> command</h3>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>samtools <span class="nb">sort</span> <span class="nt">-o</span> results/bam/SRR2584866.aligned.sorted.bam results/bam/SRR2584866.aligned.bam 
</code></pre></div>    </div>
  </blockquote>
  <blockquote class="icon">
    <h3 id="code-out-output-2"><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span> Output</h3>
    <p>Our files are pretty small, so we will not see this output. If you run the <span class="notranslate">workflow</span> with larger files, you will see something like this:</p>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">[</span>bam_sort_core] merging from 2 files...
</code></pre></div>    </div>
  </blockquote>

  <p>SAM/BAM files can be sorted in multiple ways, e.g. by location of alignment on the chromosome, by read name, etc. It is important to be aware that different alignment tools will output differently sorted SAM/BAM, and different <span class="notranslate">downstream</span> tools require differently sorted alignment files as input.</p>

  <p>You can use samtools to learn more about this bam file as well.</p>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools flagstat results/bam/SRR2584866.aligned.sorted.bam
</code></pre></div>  </div>

  <p>This will give you the following statistics about your sorted bam file:</p>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>351169 + 0 <span class="k">in </span>total <span class="o">(</span>QC-passed <span class="notranslate">reads</span> + QC-failed <span class="notranslate">reads</span><span class="o">)</span>
0 + 0 secondary
1169 + 0 supplementary
0 + 0 duplicates
351103 + 0 mapped <span class="o">(</span>99.98% : N/A<span class="o">)</span>
350000 + 0 paired <span class="k">in </span>sequencing
175000 + 0 read1
175000 + 0 read2
346688 + 0 properly paired <span class="o">(</span>99.05% : N/A<span class="o">)</span>
349876 + 0 with itself and mate mapped
58 + 0 singletons <span class="o">(</span>0.02% : N/A<span class="o">)</span>
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr <span class="o">(</span>mapQ&gt;<span class="o">=</span>5<span class="o">)</span>
</code></pre></div>  </div>
</blockquote>

<h1 class="hands_on" id="variant-calling">Variant calling</h1>

<p>A variant call is a conclusion that there is a nucleotide difference vs. some reference at a given position in an individual genome
or transcriptome, often referred to as a Single Nucleotide Variant (SNV). The call is usually accompanied by an estimate of 
variant frequency and some measure of confidence. Similar to other steps in this <span class="notranslate">workflow</span>, there are a number of tools available for 
variant calling. In this workshop we will be using <code class="language-plaintext highlighter-rouge">bcftools</code>, but there are a few things we need to do before actually calling the 
variants.</p>

<p><img src="../../images/variant_calling_workflow.png" alt="workflow. " loading="lazy" /></p>

<blockquote class="notranslate hands_on">
  <h2 id="hands_on-step-1-calculate-the-read-coverage-of-positions-in-the-genome"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Step 1: Calculate the read coverage of positions in the genome</h2>

  <p>Do the first pass on variant calling by counting read coverage with 
<a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a>. We will 
use the command <code class="language-plaintext highlighter-rouge">mpileup</code>. The flag <code class="language-plaintext highlighter-rouge">-O b</code> tells bcftools to generate a 
bcf format output file, <code class="language-plaintext highlighter-rouge">-o</code> specifies where to write the output file, and <code class="language-plaintext highlighter-rouge">-f</code> flags the path to the reference genome:</p>
  <blockquote class="icon">
    <h3 id="code-in-mpileup-command"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> <code class="language-plaintext highlighter-rouge">mpileup</code> command</h3>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>bcftools mpileup <span class="nt">-O</span> b <span class="nt">-o</span> results/bcf/SRR2584866_raw.bcf <span class="nt">-f</span> data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam 
</code></pre></div>    </div>
  </blockquote>
  <blockquote class="icon">
    <h3 id="code-out-output-3"><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span> Output</h3>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">[</span>mpileup] 1 samples <span class="k">in </span>1 input files
</code></pre></div>    </div>
  </blockquote>

  <p>We have now generated a file with coverage information for every base.</p>

</blockquote>

<blockquote class="notranslate hands_on">
  <h2 id="hands_on-step-2-detect-the-single-nucleotide-variants-snvs"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Step 2: Detect the single nucleotide variants (SNVs)</h2>

  <p>Identify SNVs using bcftools <code class="language-plaintext highlighter-rouge">call</code>. We have to specify ploidy with the flag <code class="language-plaintext highlighter-rouge">--ploidy</code>, which is one for the haploid <em>E. coli</em>. <code class="language-plaintext highlighter-rouge">-m</code> allows for multiallelic and rare-variant calling, <code class="language-plaintext highlighter-rouge">-v</code> tells the program to output variant sites only (not every site in the genome), and <code class="language-plaintext highlighter-rouge">-o</code> specifies where to write the output file:</p>
  <blockquote class="icon">
    <h3 id="code-in-call-command"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> <code class="language-plaintext highlighter-rouge">call</code> command</h3>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>bcftools call <span class="nt">--ploidy</span> 1 <span class="nt">-m</span> <span class="nt">-v</span> <span class="nt">-o</span> results/vcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf 
</code></pre></div>    </div>
  </blockquote>

</blockquote>

<blockquote class="notranslate hands_on">
  <h2 id="hands_on-step-3-filter-and-report-the-snv-variants-in-variant-calling-format-vcf"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Step 3: Filter and report the SNV variants in variant calling format (VCF)</h2>

  <p>Filter the SNVs for the final output in VCF format, using <code class="language-plaintext highlighter-rouge">vcfutils.pl</code>:</p>
  <blockquote class="icon">
    <h3 id="code-in-filtering-with-vcfutilspl-command"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> Filtering with <code class="language-plaintext highlighter-rouge">vcfutils.pl</code> command</h3>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>vcfutils.pl varFilter results/vcf/SRR2584866_variants.vcf  <span class="o">&gt;</span> results/vcf/SRR2584866_final_variants.vcf
</code></pre></div>    </div>
  </blockquote>

</blockquote>

<h1 id="hands_on-explore-the-vcf-format"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Explore the VCF format:</h1>

<blockquote class="icon">
  <p><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span></p>
  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>less <span class="nt">-S</span> results/vcf/SRR2584866_final_variants.vcf
</code></pre></div>  </div>
</blockquote>

<blockquote>
  <p>You will see the header (which describes the format), the time and date the file was
created, the version of bcftools that was used, the command line parameters used, and 
some additional information:</p>

  <p><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span> Output</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##fileformat=VCFv4.2
##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;
##bcftoolsVersion=1.8+htslib-1.8
##bcftoolsCommand=mpileup -O b -o results/bcf/SRR2584866_raw.bcf -f data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam
##reference=file://data/ref_genome/ecoli_rel606.fasta
##contig=&lt;ID=CP000819.1,length=4629812&gt;
##ALT=&lt;ID=*,Description="Represents allele(s) other than observed."&gt;
##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL."&gt;
##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description="Maximum number of <span class="notranslate">reads</span> supporting an indel"&gt;
##INFO=&lt;ID=IMF,Number=1,Type=Float,Description="Maximum fraction of <span class="notranslate">reads</span> supporting an indel"&gt;
##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Raw read depth"&gt;
##INFO=&lt;ID=VDB,Number=1,Type=Float,Description="Variant Distance Bias for filtering splice-site artefacts in <span class="notranslate">RNA</span>-seq data (bigger is better)",Version=
##INFO=&lt;ID=RPB,Number=1,Type=Float,Description="Mann-Whitney U test of Read Position Bias (bigger is better)"&gt;
##INFO=&lt;ID=MQB,Number=1,Type=Float,Description="Mann-Whitney U test of <span class="notranslate">Mapping</span> Quality Bias (bigger is better)"&gt;
##INFO=&lt;ID=BQB,Number=1,Type=Float,Description="Mann-Whitney U test of Base Quality Bias (bigger is better)"&gt;
##INFO=&lt;ID=MQSB,Number=1,Type=Float,Description="Mann-Whitney U test of <span class="notranslate">Mapping</span> Quality vs Strand Bias (bigger is better)"&gt;
##INFO=&lt;ID=SGB,Number=1,Type=Float,Description="Segregation based metric."&gt;
##INFO=&lt;ID=MQ0F,Number=1,Type=Float,Description="Fraction of MQ0 <span class="notranslate">reads</span> (smaller is better)"&gt;
##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods"&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;
##INFO=&lt;ID=ICB,Number=1,Type=Float,Description="Inbreeding Coefficient Binomial test (bigger is better)"&gt;
##INFO=&lt;ID=HOB,Number=1,Type=Float,Description="Bias in the number of HOMs number (smaller is better)"&gt;
##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes for each ALT allele, in the same order as listed"&gt;
##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;
##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases"&gt;
##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description="Average <span class="notranslate">mapping</span> quality"&gt;
##bcftools_callVersion=1.8+htslib-1.8
##bcftools_callCommand=call --ploidy 1 -m -v -o results/bcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf; Date=Tue Oct  9 18:48:10 2018
</code></pre></div>  </div>
</blockquote>

<blockquote class="icon">
  <p>Followed by information on each of the variations observed:</p>

  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  results/bam/SRR2584866.aligned.sorted.bam
CP000819.1      1521    .       C       T       207     .       DP=9;VDB=0.993024;SGB=-0.662043;MQSB=0.974597;MQ0F=0;AC=1;AN=1;DP4=0,0,4,5;MQ=60
CP000819.1      1612    .       A       G       225     .       DP=13;VDB=0.52194;SGB=-0.676189;MQSB=0.950952;MQ0F=0;AC=1;AN=1;DP4=0,0,6,5;MQ=60
CP000819.1      9092    .       A       G       225     .       DP=14;VDB=0.717543;SGB=-0.670168;MQSB=0.916482;MQ0F=0;AC=1;AN=1;DP4=0,0,7,3;MQ=60
CP000819.1      9972    .       T       G       214     .       DP=10;VDB=0.022095;SGB=-0.670168;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,8;MQ=60      GT:PL
CP000819.1      10563   .       G       A       225     .       DP=11;VDB=0.958658;SGB=-0.670168;MQSB=0.952347;MQ0F=0;AC=1;AN=1;DP4=0,0,5,5;MQ=60
CP000819.1      22257   .       C       T       127     .       DP=5;VDB=0.0765947;SGB=-0.590765;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,3;MQ=60      GT:PL
CP000819.1      38971   .       A       G       225     .       DP=14;VDB=0.872139;SGB=-0.680642;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,4,8;MQ=60      GT:PL
CP000819.1      42306   .       A       G       225     .       DP=15;VDB=0.969686;SGB=-0.686358;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,5,9;MQ=60      GT:PL
CP000819.1      45277   .       A       G       225     .       DP=15;VDB=0.470998;SGB=-0.680642;MQSB=0.95494;MQ0F=0;AC=1;AN=1;DP4=0,0,7,5;MQ=60
CP000819.1      56613   .       C       G       183     .       DP=12;VDB=0.879703;SGB=-0.676189;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,8,3;MQ=60      GT:PL
CP000819.1      62118   .       A       G       225     .       DP=19;VDB=0.414981;SGB=-0.691153;MQSB=0.906029;MQ0F=0;AC=1;AN=1;DP4=0,0,8,10;MQ=59
CP000819.1      64042   .       G       A       225     .       DP=18;VDB=0.451328;SGB=-0.689466;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,7,9;MQ=60      GT:PL
</code></pre></div>  </div>
</blockquote>

<p class="hands_on">This is a lot of information, so let‚Äôs take some time to make sure we understand our output.</p>

<p>The first few columns represent the information we have about a predicted variation.</p>

<table>
  <thead>
    <tr>
      <th>column</th>
      <th>info</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>CHROM</td>
      <td>contig location where the variation occurs</td>
    </tr>
    <tr>
      <td>POS</td>
      <td>position within the contig where the variation occurs</td>
    </tr>
    <tr>
      <td>ID</td>
      <td>a <code class="language-plaintext highlighter-rouge">.</code> until we add annotation information</td>
    </tr>
    <tr>
      <td>REF</td>
      <td>reference genotype (forward strand)</td>
    </tr>
    <tr>
      <td>ALT</td>
      <td>sample genotype (forward strand)</td>
    </tr>
    <tr>
      <td>QUAL</td>
      <td>Phred-scaled probability that the observed variant exists at this site (higher is better)</td>
    </tr>
    <tr>
      <td>FILTER</td>
      <td>a <code class="language-plaintext highlighter-rouge">.</code> if no quality filters have been applied, PASS if a filter is passed, or the name of the filters this variant failed</td>
    </tr>
  </tbody>
</table>

<p>In an ideal world, the information in the <code class="language-plaintext highlighter-rouge">QUAL</code> column would be all we needed to filter out bad variant calls.
However, in reality we need to filter on multiple other metrics.</p>

<p>The last two columns contain the genotypes and can be tricky to decode.</p>

<table>
  <thead>
    <tr>
      <th>column</th>
      <th>info</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>FORMAT</td>
      <td>lists in order the metrics presented in the final column</td>
    </tr>
    <tr>
      <td>results</td>
      <td>lists the values associated with those metrics in order</td>
    </tr>
  </tbody>
</table>

<p>For our file, the metrics presented are GT:PL:GQ.</p>

<table>
  <thead>
    <tr>
      <th>metric</th>
      <th>definition</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>AD, DP</td>
      <td>the depth per allele by sample and coverage</td>
    </tr>
    <tr>
      <td>GT</td>
      <td>the genotype for the sample at this loci. For a diploid organism, the GT field indicates the two alleles carried by the sample, encoded by a 0 for the REF allele, 1 for the first ALT allele, 2 for the second ALT allele, etc. A 0/0 means homozygous reference, 0/1 is heterozygous, and 1/1 is homozygous for the alte<span class="notranslate">rna</span>te allele.</td>
    </tr>
    <tr>
      <td>PL</td>
      <td>the likelihoods of the given genotypes</td>
    </tr>
    <tr>
      <td>GQ</td>
      <td>the Phred-scaled confidence for the genotype</td>
    </tr>
  </tbody>
</table>

<p>The Broad Institute‚Äôs <a href="https://www.broadinstitute.org/gatk/guide/article?id=1268">VCF guide</a> is an excellent place
to learn more about the VCF file format.</p>

<blockquote>
  <h2 id="question-question-1"><i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h2>

  <p>Use the <code class="language-plaintext highlighter-rouge">grep</code> and <code class="language-plaintext highlighter-rouge">wc</code> commands you have learned to assess how many variants are in the vcf file.</p>

  <blockquote>
    <h2 id="solution-solution-1"><i class="far fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h2>

    <blockquote class="icon">
      <p><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span></p>
      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ grep -v "#" results/vcf/SRR2584866_final_variants.vcf | wc -l
</code></pre></div>      </div>
    </blockquote>
    <blockquote class="icon">
      <p><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span></p>
      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>766
</code></pre></div>      </div>
    </blockquote>

    <p>There are 766 variants in this file.</p>
  </blockquote>

</blockquote>

<h1 class="question" id="hands_on-assess-the-alignment-visualization---optional-step"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Assess the alignment (visualization) - optional step</h1>

<p>It is often instructive to look at your data in a genome browser. Visualization will allow you to get a ‚Äúfeel‚Äù for 
the data, as well as detecting abnormalities and problems. Also, exploring the data in such a way may give you 
ideas for further analyses.  As such, visualization tools are useful for exploratory analysis. In this lesson we 
will describe two different tools for visualization: a light-weight command-line based one and the Broad
Institute‚Äôs Integrative Genomics Viewer (IGV) which requires
software installation and transfer of files.</p>

<blockquote class="icon">
  <p><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span>  In order for us to visualize the alignment files, we will need to index the BAM file using <code class="language-plaintext highlighter-rouge">samtools</code>:</p>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>samtools index results/bam/SRR2584866.aligned.sorted.bam
</code></pre></div>  </div>
</blockquote>
<h2 id="viewing-with-tview">Viewing with <code class="language-plaintext highlighter-rouge">tview</code></h2>

<p><a href="http://www.htslib.org/">Samtools</a> implements a very simple text alignment viewer based on the GNU
<code class="language-plaintext highlighter-rouge">ncurses</code> library, called <code class="language-plaintext highlighter-rouge">tview</code>. This alignment viewer works with short indels and shows <a href="http://maq.sourceforge.net/">MAQ</a> consensus. 
It uses different colors to display <span class="notranslate">mapping</span> quality or base quality, subjected to users‚Äô choice. Samtools viewer is known to work with a 130 GB alignment swiftly. Due to its text interface, displaying alignments over network is also very fast.</p>

<p>In order to visualize our mapped <span class="notranslate">reads</span>, we use <code class="language-plaintext highlighter-rouge">tview</code>, giving it the sorted bam file and the reference file:</p>

<h3 id="code-in-visualize-with-tview"><i class="far fa-keyboard" aria-hidden="true"></i><span class="visually-hidden">code-in</span> Visualize with <code class="language-plaintext highlighter-rouge">tview</code></h3>

<div class="language-bash icon highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>samtools tview results/bam/SRR2584866.aligned.sorted.bam data/ref_genome/ecoli_rel606.fasta
</code></pre></div></div>

<h3 id="code-out-output-4"><i class="fas fa-laptop-code" aria-hidden="true"></i><span class="visually-hidden">code-out</span> Output</h3>
<div class="language-plaintext icon highlighter-rouge"><div class="highlight"><pre class="highlight"><code>1         11        21        31        41        51        61        71        81        91        101       111       121
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATAC
..................................................................................................................................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ..................N................. ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,........................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ..................N................. ,,,,,,,,,,,,,,,,,,,,,,,,,,,.............................
...................................,g,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ....................................   ................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,....................................   ....................................      ,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ....................................  ,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,     .......
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, .............................  ,,,,,,,,,,,,,,,,,g,,,,,    ,,,,,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ...........................T.......   ,,,,,,,,,,,,,,,,,,,,,,,c,          ......
......................... ................................   ,g,,,,,,,,,,,,,,,,,,,      ...........................
,,,,,,,,,,,,,,,,,,,,, ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ,,,,,,,,,,,,,,,,,,,,,,,,,,,       ..........................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,   ................................T..  ..............................   ,,,,,,
...........................       ,,,,,,g,,,,,,,,,,,,,,,,,   ....................................         ,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,, ....................................  ...................................        ....
....................................  ........................  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,      ....
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,   ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
........................            .................................. .............................     ....
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,   ....................................        ..........................
...............................       ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ....................................
...................................  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ..................................
.................................... ,,,,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,        ,,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ............................ ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
</code></pre></div></div>

<p>The first line of output shows the genome coordinates in our reference genome. The second line shows the reference
genome sequence. The third line shows the consensus sequence determined from the sequence <span class="notranslate">reads</span>. A <code class="language-plaintext highlighter-rouge">.</code> indicates
a match to the reference sequence, so we can see that the consensus from our sample matches the reference in most
locations. That is good! If that was not the case, we should probably reconsider our choice of reference.</p>

<p>Below the horizontal line, we can see all of the <span class="notranslate">reads</span> in our sample aligned with the reference genome. Only 
positions where the called base differs from the reference are shown. You can use the arrow keys on your keyboard
to scroll or type <code class="language-plaintext highlighter-rouge">?</code> for a help menu. To navigate to a specific position, type <code class="language-plaintext highlighter-rouge">g</code>. A dialogue box will appear. In
this box, type the name of the ‚Äúchromosome‚Äù followed by a colon and the position of the variant you would like to view
(e.g. for this sample, type <code class="language-plaintext highlighter-rouge">CP000819.1:50</code> to view the 50th base. Type <code class="language-plaintext highlighter-rouge">Ctrl^C</code> or <code class="language-plaintext highlighter-rouge">q</code> to exit <code class="language-plaintext highlighter-rouge">tview</code>.</p>

<blockquote>
  <h2 id="question-question-2"><i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h2>

  <p>Visualize the alignment of the <span class="notranslate">reads</span> for our <code class="language-plaintext highlighter-rouge">SRR2584866</code> sample. What variant is present at 
position 4377265? What is the canonical nucleotide in that position?</p>

  <blockquote>
    <h2 id="solution-solution-2"><i class="far fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h2>

    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>samtools tview ~/dc_workshop/results/bam/SRR2584866.aligned.sorted.bam ~/dc_workshop/data/ref_genome/ecoli_rel606.fasta
</code></pre></div>    </div>

    <p>Then type <code class="language-plaintext highlighter-rouge">g</code>. In the dialogue box, type <code class="language-plaintext highlighter-rouge">CP000819.1:4377265</code>. 
<code class="language-plaintext highlighter-rouge">G</code> is the variant. <code class="language-plaintext highlighter-rouge">A</code> is canonical. This variant possibly changes the phenotype of this sample to hypermutable. It occurs
in the gene <em>mutL</em>, which controls <span class="notranslate">DNA</span> mismatch repair.</p>
  </blockquote>

</blockquote>

<h2 class="question" id="viewing-with-igv">Viewing with IGV</h2>

<p><a href="http://www.broadinstitute.org/igv/">IGV</a> is a stand-alone browser, which has the advantage of being installed locally and providing fast access. Web-based genome browsers, like <a href="http://www.ensembl.org/index.html">Ensembl</a> or the <a href="https://genome.ucsc.edu/">UCSC browser</a>, are slower, but provide more functionality. They not only allow for more polished and flexible visualization, but also provide easy access to a wealth of annotations and exte<span class="notranslate">rna</span>l data sources. This makes it straightforward to relate your data with information about repeat regions, known genes, epigenetic features or areas of cross-species conservation, to name just a few.</p>

<p>In order to use IGV, we will need to export some files to our local machine. We know how to do this through RStudio‚Äôs UI. 
We are going to use the down-right panel of RStudio which presents our files. We are going to need these four files:</p>
<blockquote>
  <p>~/dc_workshop/results/bam/SRR2584866.aligned.sorted.bam
~/dc_workshop/results/bam/SRR2584866.aligned.sorted.bam.bai
~/dc_workshop/data/ref_genome/ecoli_rel606.fasta
~/dc_workshop/results/vcf/SRR2584866_final_variants.vcf</p>
</blockquote>

<p>After finding them, we check the boxes left to there name. Then, after clicking the gear icon on the same panel, we choose the option ‚ÄúExport‚Äù and confirm it.
Now our files should have been downloaded locally.</p>

<p>Next, we need to open the IGV software. If you have not done so already, you can download IGV from the <a href="https://www.broadinstitute.org/software/igv/download">Broad Institute‚Äôs software page</a>, double-click the <code class="language-plaintext highlighter-rouge">.zip</code> file
to unzip it, and then drag the program into your Applications folder.</p>

<ol>
  <li>Open IGV.</li>
  <li>Load our reference genome file (<code class="language-plaintext highlighter-rouge">ecoli_rel606.fasta</code>) into IGV using the <strong>‚ÄúLoad Genomes from File‚Ä¶‚Äú</strong> option under the <strong>‚ÄúGenomes‚Äù</strong> pull-down menu.</li>
  <li>Load our BAM file (<code class="language-plaintext highlighter-rouge">SRR2584866.aligned.sorted.bam</code>) using the <strong>‚ÄúLoad from File‚Ä¶‚Äú</strong> option under the <strong>‚ÄúFile‚Äù</strong> pull-down menu.</li>
  <li>Do the same with our VCF file (<code class="language-plaintext highlighter-rouge">SRR2584866_final_variants.vcf</code>).</li>
</ol>

<p>Your IGV browser should look like the screenshot below:</p>

<p><img src="../../images/igv-screenshot.png" alt="IGV. " loading="lazy" /></p>

<p>There should be two tracks: one coresponding to our BAM file and the other for our VCF file.</p>

<p>In the <strong>VCF track</strong>, each bar across the top of the plot shows the allele fraction for a single locus. The second bar shows
the genotypes for each locus in each <em>sample</em>. We only have one sample called here, so we only see a single line. Dark blue = 
heterozygous, Cyan = homozygous variant, Grey = reference.  Filtered entries are transparent.</p>

<p>Zoom in to inspect variants you see in your filtered VCF file to become more familiar with IGV. See how quality information 
corresponds to alignment information at those loci.
Use <a href="http://software.broadinstitute.org/software/igv/AlignmentData">this website</a> and the links therein to understand how IGV colors the alignments.</p>

<blockquote class="hands_on callout">
  <h2 id="bwa-alignment-options">BWA alignment options</h2>
  <p>BWA consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM. The first algorithm is designed for Illumina sequence 
<span class="notranslate">reads</span> up to 100bp, while the other two are for sequences ranging from 70bp to 1Mbp. BWA-MEM and BWA-SW share similar features such 
as long-read support and split alignment, but BWA-MEM, which is the latest, is generally recommended for high-quality queries as it 
is faster and more accurate.</p>
</blockquote>


                
                <blockquote class="key_points">
                    <h3><i class="fas fa-key" aria-hidden="true"></i> Key points</h3>
                    <ul>
                        
                        <li><p>Bioinformatic command line tools are collections of commands that can be used to carry out bioinformatic analyses.</p>
</li>
                        
                        <li><p>To use most powerful bioinformatic tools, you will need to use the command line.</p>
</li>
                        
                        <li><p>There are many different file formats for storing genomics data. It is important to understand what type of information is contained in each file, and how it was derived.</p>
</li>
                        
                    </ul>
                </blockquote>
                

                <h1>Frequently Asked Questions</h1>
                Have questions about this tutorial? Check out the  <a href="/training-material/topics/variant-analysis/faqs/">FAQ page for the Variant Analysis topic</a> to see if your question is listed there.
                If not, please ask your question on the <a href="https://gitter.im/Galaxy-Training-Network/Lobby">GTN Gitter Channel</a> or the
                <a href="https://help.galaxyproject.org">Galaxy Help Forum</a>

                
                <h1 data-toc-skip>Useful literature</h1>
                <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/training-material/topics/variant-analysis#references">here</a>.</p>
                

                

                

                <h1>Feedback</h1>
                <p class="text-muted">Did you use this material as an instructor? Feel free to give us feedback on <a href="https://github.com/galaxyproject/training-material/issues/1452" target="_blank">how it went</a>.
                <br>Did you use this material as a learner or student? Click the form below to leave feedback.<i class="fas fa-hand-point-down"></i>
                </p>

                <div id="feedback-button">
                    <img src="/training-material/shared/images/feedback.png" title="Click to activate" alt="Click here to load Google feedback frame" />
                </div>
                <div id="feedback-form">
                </div>
                <script type="text/javascript">
                    (function (window, document) {
                        function onDocumentReady(fn) {
                            if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading") {
                                fn();
                            } else {
                                document.addEventListener('DOMContentLoaded', fn);
                            }
                        }

                        onDocumentReady(function () {
                            $("#feedback-button").click(function(evt){
                                var e = $(evt.target)
                                e.hide();

                                $("#feedback-form").html(`
                                    <iframe id="feedback-google" class="google-form" src="https://docs.google.com/forms/d/e/1FAIpQLSd4VZptFTQ03kHkMz0JyW9b6_S8geU5KjNE_tLM0dixT3ZQmA/viewform?embedded=true&entry.1235803833=Variant Calling Workflow (Variant Analysis)">Loading...</iframe>
                                `)
                            })
                        });
                    })(window, document);
                </script>



                <h1>Citing this Tutorial</h1>
                <p>
                    <ol>
                        <li id="citation-text">Sofoklis Keisaris, Fotis E. Psomopoulos, Saskia Hiltemann, Helena Rasche, The Carpentries, 2022 <b>Variant Calling Workflow (Galaxy Training Materials)</b>. <a href="https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/variant-calling/tutorial.html">https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/variant-calling/tutorial.html</a> Online; accessed TODAY
                        </li>
                        <li>
                        Batut et al., 2018 <b>Community-Driven Data Analysis Training for Biology</b> Cell Systems <a href="https://doi.org/10.1016%2Fj.cels.2018.05.012">10.1016/j.cels.2018.05.012</a>
                        </li>
                    </ol>
                </p>


                <blockquote class="details">
                  <h3><i class="fa fa-info-circle" aria-hidden="true"></i><span class="visually-hidden">details</span> BibTeX</h3>
                  <p style="display: none;">

                <div class="highlighter-rouge"><div class="highlight"><pre class="highlight">
<code id="citation-code">@misc{variant-analysis-variant-calling,
author = "Sofoklis Keisaris and Fotis E. Psomopoulos and Saskia Hiltemann and Helena Rasche and The Carpentries",
title = "Variant Calling Workflow (Galaxy Training Materials)",
year = "2022",
month = "05",
day = "18"
url = "\url{https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/variant-calling/tutorial.html}",
note = "[Online; accessed TODAY]"
}
@article{Batut_2018,
    doi = {10.1016/j.cels.2018.05.012},
    url = {https://doi.org/10.1016%2Fj.cels.2018.05.012},
    year = 2018,
    month = {jun},
    publisher = {Elsevier {BV}},
    volume = {6},
    number = {6},
    pages = {752--758.e1},
    author = {B{\'{e}}r{\'{e}}nice Batut and Saskia Hiltemann and Andrea Bagnacani and Dannon Baker and Vivek Bhardwaj and Clemens Blank and Anthony Bretaudeau and Loraine Brillet-Gu{\'{e}}guen and Martin {\v{C}}ech and John Chilton and Dave Clements and Olivia Doppelt-Azeroual and Anika Erxleben and Mallory Ann Freeberg and Simon Gladman and Youri Hoogstrate and Hans-Rudolf Hotz and Torsten Houwaart and Pratik Jagtap and Delphine Larivi{\`{e}}re and Gildas Le Corguill{\'{e}} and Thomas Manke and Fabien Mareuil and Fidel Ram{\'{\i}}rez and Devon Ryan and Florian Christoph Sigloch and Nicola Soranzo and Joachim Wolff and Pavankumar Videm and Markus Wolfien and Aisanjiang Wubuli and Dilmurat Yusuf and James Taylor and Rolf Backofen and Anton Nekrutenko and Bj√∂rn Gr√ºning},
    title = {Community-Driven Data Analysis Training for Biology},
    journal = {Cell Systems}
}</code>
                </pre></div></div>
                </p>
                </blockquote>


<script type="text/javascript">
// update the date on load, or leave fallback of 'today'
d = new Date();
document.getElementById("citation-code").innerHTML = document.getElementById("citation-code").innerHTML.replace("TODAY", d.toDateString());
document.getElementById("citation-text").innerHTML = document.getElementById("citation-text").innerHTML.replace("TODAY", d.toDateString());
</script>

                <h3><i class="far fa-thumbs-up" aria-hidden="true"></i> Congratulations on successfully completing this tutorial!</h3>

                

                

            </div>
        </div>
    </div>
</section>
<br/>
<br/>
<br/>

        </div>
        
    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap-toc.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/theme.js"></script>

    <script type="text/javascript" src="/training-material/assets/js/clipboard.min.js"></script>
    <script type="text/javascript">
    var snippets=document.querySelectorAll('div.highlight');
    [].forEach.call(snippets,function(snippet){
        snippet.firstChild.insertAdjacentHTML('beforebegin','<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets=new ClipboardJS('[data-clipboard-snippet]',{
        target:function(trigger){return trigger.nextElementSibling;
    }});
    </script>
    

    <script type="text/javascript">
        if(window.location.hostname === "galaxyproject.github.io") {
            // Redirect
            var redirect = "https://training.galaxyproject.org" + window.location.pathname + window.location.search;
            $('div.container.main-content').prepend("<div class='alert alert-warning'><strong>Note: </strong>This content has a new home at <a href=\"" + redirect + "\">" + redirect + "</a>, which you will be redirected to in 5 seconds.</div>");

            window.setTimeout(function(){
                window.location.href = redirect;
            }, 5000)

        }
    </script>
</html>
