<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>Advanced Python</title>
        
            <script async defer data-domain="training.galaxyproject.org" src="https://plausible.galaxyproject.eu/js/plausible.js"></script>

        
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap-toc.min.css">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <script src="https://kit.fontawesome.com/67b3f98409.js" crossorigin="anonymous"></script>
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />
        <link rel="alternate" type="application/atom+xml" href="/training-material/feed.xml" />

        
        
        
        
        
        <meta name="description" content="Galaxy is a scientific workflow, data integration, and da..." />
        <meta property="og:title" content="Galaxy Training: Advanced Python" />
        <meta property="og:description" content="Galaxy is a scientific workflow, data integration, and da..." />
        <meta property="og:image" content="/training-material/assets/images/GTNLogo1000.png" />
    </head>
    <body data-spy="scroll" data-target="#toc">
        <header>
    <nav class="navbar navbar-expand-md navbar-dark" aria-label="Site Navigation">
        <div class="container">
            <a class="navbar-brand" href="/training-material/">
                <img src="/training-material/assets/images/GTN-60px.png" height="30" alt="Galaxy Training Network logo">
                Galaxy Training!
            </a>

            <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#top-navbar" aria-controls="top-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        
                        <a class="nav-link" href="/training-material/topics/introduction" title="Go back to list of tutorials">
                            <i class="far fa-folder" aria-hidden="true"></i> Introduction to Galaxy Analyses
                        </a>
                        
                    </li>
                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu dropdown-menu-right">
        <!-- disable Tess for now
        <form method="get" action="https://tess.elixir-europe.org/materials">
            <input type="text" id="search" name="q" value="" style="margin-left: 0.5em;/*! border-radius: 0px; */">
            <input type="hidden" value="Galaxy Training" name="content_provider">
            <input type="submit" value="Search on TeSS" style="width: 92%;border-radius: 0px;margin: 0.5em;background: #f47d20;border: 0px;padding: 0.25em;" class="">
        </form>
        -->

        <a class="dropdown-item" href="/training-material/faq" title="Check our FAQs">
           <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> FAQs
        </a>
        
        <a class="dropdown-item" href="/training-material/topics/introduction/faqs/" title="Check our FAQs for the Introduction to Galaxy Analyses topic">
           <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Topic FAQs
        </a>
        
        <a class="dropdown-item" href="https://help.galaxyproject.org/" title="Discuss on Galaxy Help">
            <i class="far fa-comments" aria-hidden="true"></i><span class="visually-hidden">feedback</span> Galaxy Help Forum
        </a>
        <a class="dropdown-item" href="https://gitter.im/Galaxy-Training-Network/Lobby" title="Discuss on gitter">
           <i class="fab fa-gitter" aria-hidden="true"></i><span class="visually-hidden">gitter</span> Discuss on Gitter
        </a>
    </div>
</li>


                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Extras">
        <i class="far fa-star" aria-hidden="true"></i><span class="visually-hidden">galaxy-star</span> Extras
    </a>
    <div class="dropdown-menu dropdown-menu-right">

        
        <a class="dropdown-item" href="https://github.com/galaxyproject/training-material/edit/main/topics/introduction/tutorials/python-advanced-np-pd/tutorial.md" title="Edit on GitHub">
          <i class="fab fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> Edit on GitHub
        </a>

        <a class="dropdown-item" href="/training-material/stats.html" title="Show view statistics about this repository">
            <i class="fas fa-chart-bar" aria-hidden="true"></i><span class="visually-hidden">galaxy-barchart</span> GTN statistics
        </a>

        <a class="dropdown-item" href="https://plausible.galaxyproject.eu/training.galaxyproject.org?period=12mo&page=/training-material/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html" title="Show view statistics of this page">
            <i class="fas fa-chart-bar" aria-hidden="true"></i><span class="visually-hidden">galaxy-barchart</span> Page Metrics
        </a>

        
            <a class="dropdown-item" href="/training-material/feedback.html" title="Show feedback statistics about this repository">
                <i class="fas fa-chart-bar" aria-hidden="true"></i><span class="visually-hidden">galaxy-barchart</span> GTN feedback
            </a>
        

        <div class="dropdown-item">
            <div>
                <i class="fas fa-language" aria-hidden="true"></i><span class="visually-hidden">language</span> Translate this page
            </div>

            <div id="lang-selector">
                
                <strong>Automatic Translations</strong>

<a class="btn btn-info" href="https://translate.google.com/translate?hl=jp&sl=en&tl=fr&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html&edit-text=&act=url">
                Fran√ßais
                </a>
                
<a class="btn btn-info" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ja&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html&edit-text=&act=url">
                Êó•Êú¨Ë™û
                </a>
                
<a class="btn btn-info" href="https://translate.google.com/translate?hl=jp&sl=en&tl=es&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html&edit-text=&act=url">
                Espa√±ol
                </a>
                
<a class="btn btn-info" href="https://translate.google.com/translate?hl=jp&sl=en&tl=pt&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html&edit-text=&act=url">
                Portugu√™s
                </a>
                
<a class="btn btn-info" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ar&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html&edit-text=&act=url">
                ÿßŸÑÿπÿ±ÿ®Ÿäÿ©
                </a>
                
                <a class="btn btn-info" href="https://translate.google.com/translate?hl=jp&sl=en&tl=&u=https%3A%2F%2Ftraining.galaxyproject.org/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html&edit-text=&act=url" title="">
                And more!
                </a>
            </div>
        </div>

        <div class="dropdown-item">
            <div>
                <i class="fas fa-palette" aria-hidden="true"></i><span class="visually-hidden">gtn-theme</span> Theme
            </div>

            <div id="theme-selector" data-toggle="buttons">
                <label data-value="default" class="btn btn-secondary">
                    <input type="radio" name="options" id="default" autocomplete="off"> Default
                </label>
                <label data-value="night" class="btn btn-secondary">
                    <input type="radio" name="options" id="night" autocomplete="off"> Night
                </label>
                <label data-value="midnight" class="btn btn-secondary">
                    <input type="radio" name="options" id="midnight" autocomplete="off"> Midnight
                </label>
                <label data-value="rainbow" class="btn btn-secondary">
                    <input type="radio" name="options" id="rainbow" autocomplete="off"> Rainbow
                </label>
                <label data-value="progress" class="btn btn-secondary">
                    <input type="radio" name="options" id="progress" autocomplete="off">üè≥Ô∏è‚Äçüåà
                </label>
                <label data-value="halloween" class="btn btn-secondary">
                    <input type="radio" name="options" id="halloween" autocomplete="off"> üéÉ
                </label>
                <label data-value="straya" class="btn btn-secondary">
                    <input type="radio" name="options" id="downunder" autocomplete="off"> üá¶üá∫
                </label>
            </div>

        </div>

        <div class="dropdown-item">
            <div>
                <i class="fas fa-history" aria-hidden="true"></i><span class="visually-hidden">galaxy-rulebuilder-history</span> Previous Versions
            </div>

            
            <div id="archive-selector">
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/2021-09-01/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html" title="Version 2021-09-01">2021-09-01</a>
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/2021-08-01/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html" title="Version 2021-08-01">2021-08-01</a>
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/2021-07-01/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html" title="Version 2021-07-01">2021-07-01</a>
            
                <a class="btn btn-warning" href="https://training.galaxyproject.org/archive/" title="Older Versions">Older Versions</a>
            </div>

        </div>

    </div>
</li>


                    <!-- Search bar-->
                    <li class="nav-item">
                      <div id="navbarSupportedContent" role="search">
                        <!-- Search form -->
                        <form class="form-inline mr-auto" method="GET" action="/training-material/search">
                          <i class="fas fa-search nav-link" aria-hidden="true"></i>
                          <div class="md-form mb-2">
                            <input name="query" class="form-control nicer" type="text" placeholder="Search Tutorials" aria-label="Search">
                          </div>
                        </form>
                      </div>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
</header>

        <div class="container main-content" role="main">
        











<!-- Gitter -->




<script>
  ((window.gitter = {}).chat = {}).options = {
  room: 'Galaxy-Training-Network/Lobby'
  };
</script>
<script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>

<script type="application/ld+json">
    


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "citation": {
    "@type": "CreativeWork",
    "name": "Community-Driven Data Analysis Training for Biology",
    "url": "https://doi.org/10.1016/j.cels.2018.05.012"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "dateModified": "1970-01-01 00:33:41 +0000",
  "discussionUrl": "https://gitter.im/Galaxy-Training-Network/Lobby",
  "headline": "Advanced Python",
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "isFamilyFriendly": true,
  "license": "https://spdx.org/licenses/CC-BY-4.0.html",
  "producer": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "provider": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "isPartOf": {
    "@type": "CreativeWork",
    "name": "Introduction to Galaxy Analyses",
    "description": "Galaxy is a scientific workflow, data integration, and data and analysis persistence and publishing platform that aims to make computational biology accessible to research scientists that do not have computer programming experience. ",
    "url": "https://training.galaxyproject.org//training-material/topics/introduction/"
  },
  "courseCode": "introduction / python-advanced-np-pd / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for 'Advanced Python' tutorial",
  "url": "https://training.galaxyproject.org//training-material/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html",
  "timeRequired": "PT3H",
  "description": "The questions this  addresses are:\n - How can I analyze data using Python in Galaxy platform?\n\n\\nThe objectives are:\n - Use the scientific libraries pandas and numpy to explore tabular datasets\n\n",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "hasPart": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Maria Christina Maniou"
    },
    {
      "@type": "Person",
      "name": "Fotis E. Psomopoulos"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Maria Christina Maniou"
    },
    {
      "@type": "Person",
      "name": "Fotis E. Psomopoulos"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Introduction to Galaxy Analyses",
      "description": "Galaxy is a scientific workflow, data integration, and data and analysis persistence and publishing platform that aims to make computational biology accessible to research scientists that do not have computer programming experience. ",
      "url": "https://training.galaxyproject.org//training-material/topics/introduction/"
    }
  ],
  "skillLevel": "Intermediate"
}
</script>

<section class="tutorial topic-introduction">
    <h1 data-toc-skip>Advanced Python</h1>
    


    <div class="contributors-line">Authors: 
<a href="/training-material/hall-of-fame/mcmaniou/" class="contributor-badge contributor-mcmaniou"><img src="https://avatars.githubusercontent.com/mcmaniou?s=27" alt="Avatar">Maria Christina Maniou</a>


<a href="/training-material/hall-of-fame/fpsom/" class="contributor-badge contributor-fpsom"><img src="https://avatars.githubusercontent.com/fpsom?s=27" alt="Avatar">Fotis E. Psomopoulos</a>


<a href="https://training.galaxyproject.org/training-material/topics/contributing/" class="contributor-badge contributor-newcontributors"><i class="fas fa-users" aria-hidden="true"></i><span class="visually-hidden">hall-of-fame</span>  Add Contributions!</a>

</div>


    <blockquote class="overview">
        <h3>Overview</h3>
        
        <img alt="Creative Commons License" class="float-right" style="border-width:0; display: inline-block; margin:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" aria-hidden="true" />
        
        <strong><i class="far fa-question-circle" aria-hidden="true"></i> Questions:</strong>
        <ul>
        
        <li><p>How can I analyze data using Python in Galaxy platform?</p>
</li>
        
        </ul>

        <strong><i class="fas fa-bullseye" aria-hidden="true"></i> Objectives: </strong>
        <ul>
        
        <li><p>Use the scientific libraries pandas and numpy to explore tabular datasets</p>
</li>
        
        </ul>

        
        <strong><i class="fas fa-check-circle" aria-hidden="true"></i> Requirements:</strong>
        <ul>
        
        
        </ul>
        

        
        <div><strong><i class="fas fa-hourglass-half" aria-hidden="true"></i> Time estimation:</strong> 3 hours</div>
        

        
        <div><strong><i class="fas fa-graduation-cap" aria-hidden="true"></i> Level: </strong>
        


 Intermediate <span class="visually-hidden">Intermediate</span>
<span class="level intermediate" title="Intermediate Tutorial">
  <i class="fas fa-graduation-cap" aria-hidden="true"></i> <i class="fas fa-graduation-cap" aria-hidden="true"></i> <i class="fas fa-graduation-cap" aria-hidden="true"></i>
</span>

</div>
        

        
        

        
        <div id="supporting-materials"><strong><i class="fa fa-external-link" aria-hidden="true"></i> Supporting Materials:</strong></div>
        <ul class="supporting_material">
            
                <li class="btn btn-default supporting_material"><a href="/training-material/topics/introduction/slides/introduction.html" title="Topic Overview slides">
                    <i class="fab fa-slideshare" aria-hidden="true"></i> Topic Overview slides
                </a></li>
            

            
                <li class="btn btn-default supporting_material">


<a class="topic-icon" href="https://zenodo.org/record/3477564">
    <i class="far fa-copy" aria-hidden="true"></i> Datasets
</a>

</li>
            

            

            

            

            
            
            

            
                <li class="btn btn-default supporting_material">


    <a href="#" class="btn btn-default dropdown-toggle topic-icon" data-toggle="dropdown" aria-expanded="false" title="Where to run the tutorial">
        <i class="fas fa-globe" aria-hidden="true"></i><span class="visually-hidden">instances</span> Available on these Galaxies 
    </a>
    <ul class="dropdown-menu">
    
        <li>
            <a class="dropdown-item" href="https://github.com/galaxyproject/training-material/tree/main/topics/introduction/docker" title="Docker image for this tutorial">
                <i class="fab fa-docker" aria-hidden="true"></i><span class="visually-hidden">docker_image</span> Docker image
            </a>
        </li>
    
    
    </ul>


</li>
            
        </ul>
        

        <div><strong><i class="far fa-calendar" aria-hidden="true"></i> Last modification:</strong> Oct 26, 2021 </div>
        <div><strong><i class="fas fa-balance-scale" aria-hidden="true"></i> License:</strong>
            
            <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Tutorial Content is licensed under Creative Commons Attribution 4.0 International License</a>
            
            <a rel="license" href="https://github.com/galaxyproject/training-material/blob/main/LICENSE.md">The GTN Framework is licensed under MIT</a>
        </div>
    </blockquote>

    <div class="container">
        <div class="row">
            <!-- sidebar, which will move to the top on a small screen -->
            <div class="col-sm-2">
                <nav id="toc" data-toggle="toc" class="sticky-top" aria-label="Table of Contents"></nav>
            </div>
            <div class="col-sm-10">
                

                <h1 id="introduction">Introduction</h1>

<h2 id="comment">Comment</h2>
<p>This tutorial is significantly based on the Carpentries courses <a href="https://swcarpentry.github.io/python-novice-inflammation/">Programming with Python</a> and <a href="https://swcarpentry.github.io/python-novice-gapminder/">Plotting and Programming in Python</a>.</p>

<h2 id="overview">Overview</h2>
<p>In this lesson, we will be using Python 3 with some of its most popular scientific libraries. This tutorial assumes that the reader is familiar with the fundamentals of the Python programming language, as well as, how to run Python programs using <span class="notranslate">Galaxy</span>. Otherwise, it is advised to follow the ‚ÄúIntroduction to Python‚Äù tutorial available in the same platform. We will be using JupyterNotebook, a Python interpreter that comes with everything we need for the lesson. Please note:  JupyterNotebook is only currently available on the <a href="https://usegalaxy.eu/">use<span class="notranslate">galaxy</span>.eu</a> and <a href="https://usegalaxy.org/">use<span class="notranslate">galaxy</span>.org</a> sites.</p>

<h1 id="analyze-data-using-numpy">Analyze data using numpy</h1>
<p>NumPy is a python library and it stands for Numerical Python. In general, you should use this library when you want to perform operations and manipulate numerical data, especially if you have matrices or arrays. To tell Python that we‚Äôd like to start using NumPy, we need to import it:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
</code></pre></div></div>
<p>A Numpy array contains one or more elements of the same type. To examine the basic functions of the library, we will create an array of random data. These data will correspond to arthritis patients‚Äô inflammation. The rows are the individual patients, and the columns are their daily inflammation measurements. We will use the <code class="language-plaintext highlighter-rouge">random.randint()</code> function. It has 4 arguments as inputs <code class="language-plaintext highlighter-rouge">randint(low, high=None, size=None, dtype=int)</code>. <code class="language-plaintext highlighter-rouge">low</code> nad <code class="language-plaintext highlighter-rouge">high</code> specify the limits of the random number generator. <code class="language-plaintext highlighter-rouge">size</code> determines the shape of the array and it can be an integer or a tuple.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">np</span><span class="p">.</span><span class="n">random</span><span class="p">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">2021</span><span class="p">)</span>  <span class="c1">#create reproducible work
</span><span class="n">random_data</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">random</span><span class="p">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span><span class="mi">70</span><span class="p">))</span>
</code></pre></div></div>
<p>If we want to check the data have been loaded, we can print the variable‚Äôs value:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">random_data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">[[21 22  1 ...  8 21 12]
 [20 19 15 ...  5 10  7]
 [ 7 24 14 ... 24 19 12]
 ...
 [22 16  3 ...  2 18 21]
 [22 14 23 ...  7 13  4]
 [ 7  7  8 ... 15  8  9]]
</code></pre>
<p>Now that the data are in memory, we can manipulate them. First, let‚Äôs ask what type of thing data refers to:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">random_data</span><span class="p">))</span>
</code></pre></div></div>
<pre><code class="language-Output">&lt;class 'numpy.ndarray'&gt;
</code></pre>

<p>The output tells us that data currently refers to an N-dimensional array, the functionality for which is provided by the NumPy library. These data correspond to arthritis patients‚Äô inflammation. The rows are the individual patients, and the columns are their daily inflammation measurements.</p>

<p>The <code class="language-plaintext highlighter-rouge">type</code> function will only tell you that a variable is a NumPy array but won‚Äôt tell you the type of thing inside the array. We can find out the type of the data contained in the NumPy array.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">random_data</span><span class="p">.</span><span class="n">dtype</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-Output">int64
</code></pre>
<p>This tells us that the NumPy array‚Äôs elements are integer numbers.</p>

<p>With the following command, we can see the array‚Äôs shape:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">random_data</span><span class="p">.</span><span class="n">shape</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-Output">(50, 70)
</code></pre>
<p>The output tells us that the data array variable contains 50 rows and 70 columns. When we created the variable <code class="language-plaintext highlighter-rouge">random_data</code> to store our arthritis data, we did not only create the array; we also created information about the array, called members or attributes. This extra information describes <code class="language-plaintext highlighter-rouge">random_data</code> in the same way an adjective describes a noun. <code class="language-plaintext highlighter-rouge">random_data.shape</code> is an attribute of <code class="language-plaintext highlighter-rouge">random_data</code> which describes the dimensions of <code class="language-plaintext highlighter-rouge">random_data</code>.</p>

<p>If we want to get a single number from the array, we must provide an index in square brackets after the variable name, just as we do in math when referring to an element of a matrix. Our data has two dimensions, so we will need to use two indices to refer to one specific value:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="s">'first value in data:'</span><span class="p">,</span> <span class="n">random_data</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
</code></pre></div></div>
<pre><code class="language-output">first value in data: 21
</code></pre>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="s">'middle value in data:'</span><span class="p">,</span> <span class="n">random_data</span><span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">35</span><span class="p">])</span>
</code></pre></div></div>
<pre><code class="language-output">middle value in data: 11
</code></pre>

<p>The <span class="notranslate">expression</span> random_data[25, 35] accesses the element at row 25, column 35. While this <span class="notranslate">expression</span> may not surprise you, random_data[0, 0] might. Programming languages like Fortran, MATLAB and R start counting at 1 because that‚Äôs what human beings have done for thousands of years. Languages in the C family (including C++, Java, Perl, and Python) count from 0 because it represents an offset from the first value in the array (the second value is offset by one index from the first value). As a result, if we have an M√óN array in Python, its indices go from 0 to M-1 on the first axis and 0 to N-1 on the second.</p>

<p>Slicing data
An index like [25, 35] selects a single element of an array, but we can select whole sections as well, using slicing the same way as previously with the strings. For example, we can select the first ten days (columns) of values for the first four patients (rows) like this:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">random_data</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>
</code></pre></div></div>
<pre><code class="language-output">[[21 22  1 14 23 13 22 13 23  7]
 [20 19 15 19 22 19  1 21 24  4]
 [ 7 24 14 11  5  6  3  4  3 17]
 [24  9  4 17 10  4 11  4 13 20]]
</code></pre>

<p>We don‚Äôt have to include the upper and lower bound on the slice. If we don‚Äôt include the lower bound, Python uses 0 by default; if we don‚Äôt include the upper, the slice runs to the end of the axis, and if we don‚Äôt include either (i.e., if we use ‚Äò:‚Äô on its own), the slice includes everything:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">small</span> <span class="o">=</span> <span class="n">random_data</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">36</span><span class="p">:]</span>
<span class="k">print</span><span class="p">(</span><span class="s">'small is:'</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">small</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">small is:
[[22 23 19 11  5 14 11 20 19 16  9 11 13 22 12 20  2 15 18 19  8 22 20 13
  24 20 10  5 10  8  8  8 21 12]
 [ 7 15 12 10 20  2 12 20 16  7 18 12  1  6 12 17 24  1 18  4  7 13 12 21
   8 10  5  2  8 18 16  5 10  7]
 [23 23  8  5 20 14  3  9 17 13 22  9  2 11 14 23  7 20  8 22  5 21 20  2
  11  4 24 19  2 14 22 24 19 12]]
</code></pre>

<p>The above example selects rows 0 through 2 and columns 36 through to the end of the array.</p>

<h2 id="process-the-data">Process the data</h2>
<p>NumPy has several useful functions that take an array as input to perform operations on its values. If we want to find the average inflammation for all patients on all days, for example, we can ask NumPy to compute random_data‚Äôs mean value:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="n">mean</span><span class="p">(</span><span class="n">random_data</span><span class="p">))</span>
</code></pre></div></div>
<pre><code class="language-output">12.567142857142857
</code></pre>

<p>Let‚Äôs use three other NumPy functions to get some descriptive values about the dataset. We‚Äôll also use multiple assignment, a convenient Python feature that will enable us to do this all in one line.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">maxval</span><span class="p">,</span> <span class="n">minval</span><span class="p">,</span> <span class="n">stdval</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="nb">max</span><span class="p">(</span><span class="n">random_data</span><span class="p">),</span> <span class="n">np</span><span class="p">.</span><span class="nb">min</span><span class="p">(</span><span class="n">random_data</span><span class="p">),</span> <span class="n">np</span><span class="p">.</span><span class="n">std</span><span class="p">(</span><span class="n">random_data</span><span class="p">)</span>

<span class="k">print</span><span class="p">(</span><span class="s">'maximum inflammation:'</span><span class="p">,</span> <span class="n">maxval</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s">'minimum inflammation:'</span><span class="p">,</span> <span class="n">minval</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s">'standard deviation:'</span><span class="p">,</span> <span class="n">stdval</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">maximum inflammation: 24
minimum inflammation: 1
standard deviation: 6.966639309258527
</code></pre>

<p>How did we know what functions NumPy has and how to use them? If you are working in IPython or in a Jupyter Notebook, there is an easy way to find out. If you type the name of something followed by a dot, then you can use tab completion (e.g. type <code class="language-plaintext highlighter-rouge">np.</code> and then press Tab) to see a list of all functions and attributes that you can use. After selecting one, you can also add a question mark (e.g. <code class="language-plaintext highlighter-rouge">np.cumprod?</code>), and IPython will return an explanation of the method! This is the same as doing <code class="language-plaintext highlighter-rouge">help(np.cumprod)</code>.</p>

<p>When analyzing data, though, we often want to look at variations in statistical values, such as the maximum inflammation per patient or the average inflammation per day. One way to do this is to create a new temporary array of the data we want, then ask it to do the calculation:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">patient_0</span> <span class="o">=</span> <span class="n">random_data</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span> <span class="c1"># 0 on the first axis (rows), everything on the second (columns)
</span><span class="k">print</span><span class="p">(</span><span class="s">'maximum inflammation for patient 0:'</span><span class="p">,</span> <span class="n">np</span><span class="p">.</span><span class="nb">max</span><span class="p">(</span><span class="n">patient_0</span><span class="p">))</span>
</code></pre></div></div>
<pre><code class="language-output">maximum inflammation for patient 0: 24
</code></pre>

<p>What if we need the maximum inflammation for each patient over all days (as in the next diagram on the left) or the average for each day (as in the diagram on the right)? As the diagram below shows, we want to perform the operation across an axis:</p>

<p><img src="../../images/python-advanced-np-pd/Figure8_Operations_across_axis.png" alt="Figure 8_Operations_across_axis. " loading="lazy" /></p>

<p>To support this functionality, most array functions allow us to specify the axis we want to work on. If we ask for the average across axis 0 (rows in our 2D example), we get:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="n">mean</span><span class="p">(</span><span class="n">random_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
</code></pre></div></div>
<pre><code class="language-output">[12.9  13.4  12.66 12.54 12.76 12.22 10.84 12.32 12.56 13.5  11.36 14.62
 12.18 13.02 13.3  10.6  11.36 10.26 10.22 13.14 14.12 12.5  12.3  12.78
 12.4  12.08 13.42 12.64 11.5  12.08 14.28 13.2  12.48 13.4  12.54 13.84
 12.88 13.74 10.88 11.38 11.7  11.7  12.62 13.8  14.02 12.92 12.98 12.5
 11.52 13.96 11.98 14.5  12.16 11.04 13.96 12.48 11.98 12.32 12.2  13.88
 10.96 13.44 12.26 12.98 11.46 12.52 13.9  13.   12.4  12.36]
</code></pre>

<p>As a quick check, we can ask this array what its shape is:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="n">mean</span><span class="p">(</span><span class="n">random_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">).</span><span class="n">shape</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-Output">(70,)
</code></pre>

<p>The <span class="notranslate">expression</span> (70,) tells us we have an N√ó1 vector, so this is the average inflammation per day for all patients. If we average across axis 1 (columns in our 2D example), we get the average inflammation per patient across all days.:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="n">mean</span><span class="p">(</span><span class="n">random_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
</code></pre></div></div>
<pre><code class="language-output">[13.28571429 12.37142857 13.31428571 12.21428571 14.24285714 13.58571429
 11.55714286 13.02857143 12.44285714 12.37142857 13.15714286 12.32857143
 11.37142857 11.94285714 12.37142857 11.81428571 12.48571429 12.47142857
 13.14285714 13.05714286 12.4        13.7        13.28571429 13.01428571
 12.8        12.05714286 11.57142857 12.25714286 12.62857143 12.12857143
 13.94285714 11.88571429 12.02857143 12.98571429 12.15714286 12.24285714
 12.48571429 13.87142857 12.27142857 13.77142857 12.04285714 12.34285714
 13.58571429 13.04285714 12.28571429 11.35714286 10.3        12.72857143
 11.81428571 12.81428571]
</code></pre>

<h3 id="stacking-arrays">Stacking arrays</h3>
<p>Arrays can be concatenated and stacked on top of one another, using NumPy‚Äôs <code class="language-plaintext highlighter-rouge">vstack</code> and <code class="language-plaintext highlighter-rouge">hstack</code> functions for vertical and horizontal stacking, respectively.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>

<span class="n">A</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">]])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'A = '</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">A</span><span class="p">)</span>

<span class="n">B</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">A</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'B = '</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">B</span><span class="p">)</span>

<span class="n">C</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">A</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'C = '</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">C</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-Output">A =
[[1 2 3]
 [4 5 6]
 [7 8 9]]
B =
[[1 2 3 1 2 3]
 [4 5 6 4 5 6]
 [7 8 9 7 8 9]]
C =
[[1 2 3]
 [4 5 6]
 [7 8 9]
 [1 2 3]
 [4 5 6]
 [7 8 9]]
</code></pre>

<h3 id="remove-nan-values">Remove NaN values</h3>

<p>Sometimes there are missing values in an array, that could make it difficult to perform operations on it. To remove the <code class="language-plaintext highlighter-rouge">NaN</code> you must first find their indexes and then replace them. The following example replaces them with <code class="language-plaintext highlighter-rouge">0</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">a</span> <span class="o">=</span> <span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">NaN</span><span class="p">]])</span>
<span class="k">print</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="n">a</span><span class="p">[</span><span class="n">np</span><span class="p">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">a</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">print</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-Output">[[ 1.  2.  3.]
 [ 0.  3. nan]]
 
[[1. 2. 3.]
 [0. 3. 0.]]
</code></pre>

<p>### Exercises</p>

<ol>
  <li>
    <p>Write some additional code that slices the first and last columns of A, and stacks them into a 3x2 array. Make sure to print the results to verify your solution.</p>
  </li>
  <li>
    <p>Given the followind array <code class="language-plaintext highlighter-rouge">A</code>, keep only the elements that are lower that <code class="language-plaintext highlighter-rouge">0.05</code>.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">A</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.81</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.67</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">])</span>
</code></pre></div>    </div>
  </li>
</ol>

<p>### Solutions</p>
<ol>
  <li>
    <p>A ‚Äògotcha‚Äô with array indexing is that singleton dimensions are dropped by default. That means <code class="language-plaintext highlighter-rouge">A[:, 0]</code> is a one dimensional array, which won‚Äôt stack as desired. To preserve singleton dimensions, the index itself can be a slice or array. For example, <code class="language-plaintext highlighter-rouge">A[:, :1]</code> returns a two dimensional array with one singleton dimension (i.e. a column vector).</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">D</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">A</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">1</span><span class="p">],</span> <span class="n">A</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">:]))</span>
<span class="k">print</span><span class="p">(</span><span class="s">'D = '</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
</code></pre></div>    </div>
    <pre><code class="language-Output">D =
[[1 3]
 [4 6]
 [7 9]]
</code></pre>
  </li>
  <li>
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">A</span> <span class="o">=</span> <span class="n">A</span><span class="p">[</span><span class="n">A</span><span class="o">&lt;</span><span class="mf">0.05</span><span class="p">]</span>
</code></pre></div>    </div>
  </li>
</ol>

<h1 id="use-pandas-to-work-with-dataframes">Use pandas to work with dataframes</h1>

<p>Pandas is a widely-used Python library for statistics, particularly on tabular data. If you are familiar with R dataframes, then this is the library that integrates this functionality. A dataframe is a 2-dimensional table with indexes and column names. The indexes indicate the difference in rows, while the column names indicate the difference in columns. You will see later that these two features are useful when you‚Äôre manipulating your data. Each column can contain different data types.</p>

<p>Load it with import pandas as <code class="language-plaintext highlighter-rouge">pd</code>. The alias <code class="language-plaintext highlighter-rouge">pd</code> is commonly used for pandas.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
</code></pre></div></div>

<p>There are many ways to create a pandas dataframe. For example you can use a numpy array as input.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([[</span><span class="s">''</span><span class="p">,</span><span class="s">'Col1'</span><span class="p">,</span><span class="s">'Col2'</span><span class="p">],</span>
                <span class="p">[</span><span class="s">'Row1'</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span>
                <span class="p">[</span><span class="s">'Row2'</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]])</span>
                
<span class="k">print</span><span class="p">(</span><span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span><span class="mi">1</span><span class="p">:],</span>
                  <span class="n">index</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span><span class="mi">0</span><span class="p">],</span>
                  <span class="n">columns</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:]))</span>
</code></pre></div></div>
<pre><code class="language-output">     Col1 Col2
Row1    1    2
Row2    3    4
</code></pre>

<p>For the purposes of this tutorial, we will use a file with the annotated differentially expressed genes that was produced in the <a href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/ref-based/tutorial.html">Reference-based <span class="notranslate">RNA</span>-Seq data analysis</a> tutorial.</p>

<p>We can read a tabular file with <code class="language-plaintext highlighter-rouge">pd.read_csv</code>. The first argument is the filepath of the file to be read. The <code class="language-plaintext highlighter-rouge">sep</code> argument refers to the symbol used to separate the data into different columns. You can check the rest of the arguments using the <code class="language-plaintext highlighter-rouge">help()</code> function.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://zenodo.org/record/3477564/files/annotatedDEgenes.tabular"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">          GeneID     Base mean  log2(FC)    StdErr  Wald-Stats        P-value  \
0    FBgn0039155   1086.974295 -4.148450  0.134949  -30.740913  1.617357e-207   
1    FBgn0003360   6409.577128 -2.999777  0.104345  -28.748637  9.419922e-182   
2    FBgn0026562  65114.840564 -2.380164  0.084327  -28.225437  2.850430e-175   
3    FBgn0025111   2192.322369  2.699939  0.097945   27.565978  2.847517e-167   
4    FBgn0029167   5430.067277 -2.105062  0.092547  -22.745964  1.573284e-114   
..           ...           ...       ...       ...         ...            ...   
125  FBgn0035710     26.161771  1.048979  0.232922    4.503559   6.682473e-06   
126  FBgn0035523     70.998197  1.004819  0.223763    4.490561   7.103599e-06   
127  FBgn0038261     44.270577  1.006264  0.224124    4.489756   7.130487e-06   
128  FBgn0039178     23.550056  1.040917  0.232626    4.474631   7.654328e-06   
129  FBgn0034636     24.770519 -1.028531  0.232168   -4.430118   9.418135e-06   

             P-adj Chromosome     Start       End Strand         Feature  \
0    1.387207e-203      chr3R  24141394  24147490      +  protein_coding   
1    4.039734e-178       chrX  10780892  10786958      -  protein_coding   
2    8.149380e-172      chr3R  26869237  26871995      -  protein_coding   
3    6.105789e-164       chrX  10778953  10786907      -  protein_coding   
4    2.698811e-111      chr3L  13846053  13860001      +  protein_coding   
..             ...        ...       ...       ...    ...             ...   
125   1.436480e-04      chr3L   6689326   6703521      -  protein_coding   
126   1.523189e-04      chr3L   4277961   4281585      +  protein_coding   
127   1.525142e-04      chr3R  14798985  14801163      +  protein_coding   
128   1.629061e-04      chr3R  24283954  24288617      +  protein_coding   
129   1.951192e-04      chr2R  21560245  21576035      -  protein_coding   

       Gene name  
0           Kal1  
1           sesB  
2    BM-40-SPARC  
3           Ant2  
4            Hml  
..           ...  
125       SP1173  
126       CG1311  
127      CG14856  
128       CG6356  
129      CG10440  

[130 rows x 13 columns]
</code></pre>
<p>The columns in a dataframe are the observed variables, and the rows are the observations. Pandas uses backslash <code class="language-plaintext highlighter-rouge">\</code> to show wrapped lines when output is too wide to fit the screen.</p>

<h2 id="explore-the-data">Explore the data</h2>

<p>You can use <code class="language-plaintext highlighter-rouge">index_col</code> to specify that a column‚Äôs values should be used as row headings.</p>

<p>By default row indexes are numbers, but we could use a column of the data. To pass the name of the column to <code class="language-plaintext highlighter-rouge">read_csv</code>, you can use its <code class="language-plaintext highlighter-rouge">index_col</code> parameter. Be careful though, because the row indexes must be unique for each row.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://zenodo.org/record/3477564/files/annotatedDEgenes.tabular"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="s">'GeneID'</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">                Base mean  log2(FC)    StdErr  Wald-Stats        P-value  \
GeneID                                                                     
FBgn0039155   1086.974295 -4.148450  0.134949  -30.740913  1.617357e-207   
FBgn0003360   6409.577128 -2.999777  0.104345  -28.748637  9.419922e-182   
FBgn0026562  65114.840564 -2.380164  0.084327  -28.225437  2.850430e-175   
FBgn0025111   2192.322369  2.699939  0.097945   27.565978  2.847517e-167   
FBgn0029167   5430.067277 -2.105062  0.092547  -22.745964  1.573284e-114   
...                   ...       ...       ...         ...            ...   
FBgn0035710     26.161771  1.048979  0.232922    4.503559   6.682473e-06   
FBgn0035523     70.998197  1.004819  0.223763    4.490561   7.103599e-06   
FBgn0038261     44.270577  1.006264  0.224124    4.489756   7.130487e-06   
FBgn0039178     23.550056  1.040917  0.232626    4.474631   7.654328e-06   
FBgn0034636     24.770519 -1.028531  0.232168   -4.430118   9.418135e-06   

                     P-adj Chromosome     Start       End Strand  \
GeneID                                                             
FBgn0039155  1.387207e-203      chr3R  24141394  24147490      +   
FBgn0003360  4.039734e-178       chrX  10780892  10786958      -   
FBgn0026562  8.149380e-172      chr3R  26869237  26871995      -   
FBgn0025111  6.105789e-164       chrX  10778953  10786907      -   
FBgn0029167  2.698811e-111      chr3L  13846053  13860001      +   
...                    ...        ...       ...       ...    ...   
FBgn0035710   1.436480e-04      chr3L   6689326   6703521      -   
FBgn0035523   1.523189e-04      chr3L   4277961   4281585      +   
FBgn0038261   1.525142e-04      chr3R  14798985  14801163      +   
FBgn0039178   1.629061e-04      chr3R  24283954  24288617      +   
FBgn0034636   1.951192e-04      chr2R  21560245  21576035      -   

                    Feature    Gene name  
GeneID                                    
FBgn0039155  protein_coding         Kal1  
FBgn0003360  protein_coding         sesB  
FBgn0026562  protein_coding  BM-40-SPARC  
FBgn0025111  protein_coding         Ant2  
FBgn0029167  protein_coding          Hml  
...                     ...          ...  
FBgn0035710  protein_coding       SP1173  
FBgn0035523  protein_coding       CG1311  
FBgn0038261  protein_coding      CG14856  
FBgn0039178  protein_coding       CG6356  
FBgn0034636  protein_coding      CG10440  

[130 rows x 12 columns]
</code></pre>

<p>You can use the <code class="language-plaintext highlighter-rouge">DataFrame.info()</code> method to find out more about a dataframe.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span><span class="p">.</span><span class="n">info</span><span class="p">()</span>
</code></pre></div></div>
<pre><code class="language-output">&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 130 entries, FBgn0039155 to FBgn0034636
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Base mean   130 non-null    float64
 1   log2(FC)    130 non-null    float64
 2   StdErr      130 non-null    float64
 3   Wald-Stats  130 non-null    float64
 4   P-value     130 non-null    float64
 5   P-adj       130 non-null    float64
 6   Chromosome  130 non-null    object 
 7   Start       130 non-null    int64  
 8   End         130 non-null    int64  
 9   Strand      130 non-null    object 
 10  Feature     130 non-null    object 
 11  Gene name   130 non-null    object 
dtypes: float64(6), int64(2), object(4)
memory usage: 13.2+ KB
</code></pre>

<p>We learn that this is a DataFrame. It consists of 130 rows and 12 columns. None of the columns contains any missing values. 6 columns contain 64-bit floating point <code class="language-plaintext highlighter-rouge">float64</code> values, 2 contain 64-bit integer <code class="language-plaintext highlighter-rouge">int64</code> values and 4 contain character <code class="language-plaintext highlighter-rouge">object</code> values. It uses 13.2KB of memory.</p>

<p>The <code class="language-plaintext highlighter-rouge">DataFrame.columns</code> variable stores information about the dataframe‚Äôs columns.</p>

<p>Note that this is an attribute, not a method. (It doesn‚Äôt have parentheses.) Called a member variable, or just member.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">columns</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">Index(['Base mean', 'log2(FC)', 'StdErr', 'Wald-Stats', 'P-value', 
	   'P-adj', 'Chromosome', 'Start', 'End', 'Strand', 'Feature', 'Gene name'],
      dtype='object')
</code></pre>

<p>You could use <code class="language-plaintext highlighter-rouge">DataFrame.T</code> to transpose a dataframe. The <code class="language-plaintext highlighter-rouge">Transpose</code> (written <code class="language-plaintext highlighter-rouge">.T</code>) doesn‚Äôt copy the data, just changes the program‚Äôs view of it. Like columns, it is a member variable.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">T</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">GeneID         FBgn0039155     FBgn0003360     FBgn0026562     FBgn0025111  \
Base mean      1086.974295     6409.577128    65114.840564     2192.322369   
log2(FC)          -4.14845       -2.999777       -2.380164        2.699939   
StdErr            0.134949        0.104345        0.084327        0.097945   
Wald-Stats      -30.740913      -28.748637      -28.225437       27.565978   
P-value                0.0             0.0             0.0             0.0   
P-adj                  0.0             0.0             0.0             0.0   
Chromosome           chr3R            chrX           chr3R            chrX   
Start             24141394        10780892        26869237        10778953   
End               24147490        10786958        26871995        10786907   
Strand                   +               -               -               -   
Feature     protein_coding  protein_coding  protein_coding  protein_coding   
Gene name             Kal1            sesB     BM-40-SPARC            Ant2   

GeneID         FBgn0029167     FBgn0039827     FBgn0035085     FBgn0034736  \
Base mean      5430.067277      390.901782      928.263812      330.383023   
log2(FC)         -2.105062       -3.503013       -2.414074       -3.018179   
StdErr            0.092547         0.16003        0.115185        0.158154   
Wald-Stats      -22.745964      -21.889756      -20.958204      -19.083791   
P-value                0.0             0.0             0.0             0.0   
P-adj                  0.0             0.0             0.0             0.0   
Chromosome           chr3L           chr3R           chr2R           chr2R   
Start             13846053        31196915        24945138        22550093   
End               13860001        31203722        24946636        22552113   
Strand                   +               +               +               +   
Feature     protein_coding  protein_coding  protein_coding  protein_coding   
Gene name              Hml          CG1544          CG3770          CG6018   

GeneID     FBgn0264475     FBgn0000071  ...     FBgn0264343     FBgn0038237  \
Base mean   955.454454      468.057926  ...       57.156629       43.409588   
log2(FC)     -2.334486        2.360017  ...        1.036055       -1.105893   
StdErr         0.12423        0.135644  ...        0.216472        0.232871   
Wald-Stats  -18.791643       17.398617  ...        4.786091       -4.748945   
P-value            0.0             0.0  ...        0.000002        0.000002   
P-adj              0.0             0.0  ...        0.000042        0.000049   
Chromosome       chr3L           chr3R  ...           chr2L           chr3R   
Start           820758         6762592  ...         7227733        14513900   
End             821512         6765261  ...         7228641        14558304   
Strand               +               +  ...               +               -   
Feature        linc<span class="notranslate">RNA</span>  protein_coding  ...  protein_coding  protein_coding   
Gene name      CR43883             Ama  ...         CG43799            Pde6   

GeneID         FBgn0020376     FBgn0028939     FBgn0036560     FBgn0035710  \
Base mean         61.54154       31.587685       27.714241       26.161771   
log2(FC)         -1.038158       -1.091024           1.089        1.048979   
StdErr            0.219867        0.232324        0.232753        0.232922   
Wald-Stats       -4.721761       -4.696136        4.678786        4.503559   
P-value           0.000002        0.000003        0.000003        0.000007   
P-adj             0.000055        0.000062        0.000067        0.000144   
Chromosome           chr2L           chr2L           chr3L           chr3L   
Start              4120342        13980125        16035484         6689326   
End                4121627        13983269        16037227         6703521   
Strand                   +               +               +               -   
Feature     protein_coding  protein_coding  protein_coding  protein_coding   
Gene name          Sr-CIII           NimC2          CG5895          SP1173   

GeneID         FBgn0035523     FBgn0038261     FBgn0039178     FBgn0034636  
Base mean        70.998197       44.270577       23.550056       24.770519  
log2(FC)          1.004819        1.006264        1.040917       -1.028531  
StdErr            0.223763        0.224124        0.232626        0.232168  
Wald-Stats        4.490561        4.489756        4.474631       -4.430118  
P-value           0.000007        0.000007        0.000008        0.000009  
P-adj             0.000152        0.000153        0.000163        0.000195  
Chromosome           chr3L           chr3R           chr3R           chr2R  
Start              4277961        14798985        24283954        21560245  
End                4281585        14801163        24288617        21576035  
Strand                   +               +               +               -  
Feature     protein_coding  protein_coding  protein_coding  protein_coding  
Gene name           CG1311         CG14856          CG6356         CG10440  

[12 rows x 130 columns]
</code></pre>

<p>You can use <code class="language-plaintext highlighter-rouge">DataFrame.describe()</code> to get summary statistics about the data. <code class="language-plaintext highlighter-rouge">DataFrame.describe()</code> returns the summary statistics of only the columns that have numerical data.  All other columns are ignored, unless you use the argument <code class="language-plaintext highlighter-rouge">include='all'</code>. Depending on the data type of each column, the statistics that can‚Äôt be calculated are replaced with  the value <code class="language-plaintext highlighter-rouge">NaN</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">describe</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="s">'all'</span><span class="p">))</span>
</code></pre></div></div>
<pre><code class="language-output">           Base mean    log2(FC)      StdErr  Wald-Stats        P-value  \
count     130.000000  130.000000  130.000000  130.000000   1.300000e+02   
unique           NaN         NaN         NaN         NaN            NaN   
top              NaN         NaN         NaN         NaN            NaN   
freq             NaN         NaN         NaN         NaN            NaN   
mean     1911.266780   -0.207426    0.164324   -1.901479   4.200552e-07   
std      6888.074171    1.612615    0.042358   11.037352   1.531253e-06   
min        19.150759   -4.148450    0.084327  -30.740913  1.617357e-207   
25%       100.286324   -1.336283    0.128488  -10.004594   4.952408e-31   
50%       237.986359   -1.027150    0.163704   -4.981688   2.109219e-19   
75%       948.656793    1.220304    0.198655    7.692073   6.731953e-12   
max     65114.840564    2.699939    0.232922   27.565978   9.418135e-06   

                P-adj Chromosome         Start           End Strand  \
count    1.300000e+02        130  1.300000e+02  1.300000e+02    130   
unique            NaN          5           NaN           NaN      2   
top               NaN      chr3R           NaN           NaN      +   
freq              NaN         32           NaN           NaN     72   
mean     9.321175e-06        NaN  1.343684e+07  1.344644e+07    NaN   
std      3.278527e-05        NaN  7.970664e+06  7.973420e+06    NaN   
min     1.387207e-203        NaN  1.274480e+05  1.403400e+05    NaN   
25%      1.275950e-28        NaN  7.277516e+06  7.279063e+06    NaN   
50%      2.563262e-17        NaN  1.316155e+07  1.316625e+07    NaN   
75%      3.706020e-10        NaN  1.925043e+07  1.928464e+07    NaN   
max      1.951192e-04        NaN  3.119692e+07  3.120372e+07    NaN   

               Feature Gene name  
count              130       130  
unique               3       130  
top     protein_coding      Sesn  
freq               126         1  
mean               NaN       NaN  
std                NaN       NaN  
min                NaN       NaN  
25%                NaN       NaN  
50%                NaN       NaN  
75%                NaN       NaN  
max                NaN       NaN  
</code></pre>

<h3 id="exercises">Exercises</h3>

<ol>
  <li>
    <p>After reading the data, use <code class="language-plaintext highlighter-rouge">help(data.head)</code> and <code class="language-plaintext highlighter-rouge">help(data.tail)</code> to find out what <code class="language-plaintext highlighter-rouge">DataFrame.head</code> and <code class="language-plaintext highlighter-rouge">DataFrame.tail</code> do.
a. What method call will display the first three rows of the data?
b. What method call will display the last three columns of this data? (Hint: you may need to change your view of the data.)</p>
  </li>
  <li>
    <p>As well as the <code class="language-plaintext highlighter-rouge">read_csv</code> function for reading data from a file, Pandas provides a <code class="language-plaintext highlighter-rouge">to_csv</code> function to write dataframes to files. Applying what you‚Äôve learned about reading from files, write one of your dataframes to a file called <code class="language-plaintext highlighter-rouge">processed.csv</code>. You can use <code class="language-plaintext highlighter-rouge">help</code> to get information on how to use <code class="language-plaintext highlighter-rouge">to_csv</code>.</p>
  </li>
</ol>

<h3 id="solutions">Solutions</h3>

<ol>
  <li>
    <p>a. We can check out the first five rows of the data by executing <code class="language-plaintext highlighter-rouge">data.head()</code> (allowing us to view the head of the DataFrame). We can specify the number of rows we wish to see by specifying the parameter <code class="language-plaintext highlighter-rouge">n</code> in our call to <code class="language-plaintext highlighter-rouge">data.head()</code>. To view the first three rows, execute:</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">data</span><span class="p">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div>    </div>
  </li>
</ol>

<table>
  <thead>
    <tr>
      <th>¬†</th>
      <th>Base mean</th>
      <th>log2(FC)</th>
      <th>StdErr</th>
      <th>Wald-Stats</th>
      <th>P-value</th>
      <th>P-adj</th>
      <th>Chromosome</th>
      <th>Start</th>
      <th>End</th>
      <th>Strand</th>
      <th>Feature</th>
      <th>Gene name</th>
      <th>GeneID</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>FBgn0039155</td>
      <td>1086.974295</td>
      <td>-4.148450</td>
      <td>0.134949</td>
      <td>-30.740913</td>
      <td>1.617357e-207</td>
      <td>1.387207e-203</td>
      <td>chr3R</td>
      <td>24141394</td>
      <td>24147490</td>
      <td>+</td>
      <td>protein_coding</td>
      <td>Kal1</td>
      <td>¬†</td>
    </tr>
    <tr>
      <td>FBgn0003360</td>
      <td>6409.577128</td>
      <td>-2.999777</td>
      <td>0.104345</td>
      <td>-28.748637</td>
      <td>9.419922e-182</td>
      <td>4.039734e-178</td>
      <td>chrX</td>
      <td>10780892</td>
      <td>10786958</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>sesB</td>
      <td>¬†</td>
    </tr>
    <tr>
      <td>FBgn0026562</td>
      <td>65114.840564</td>
      <td>-2.380164</td>
      <td>0.084327</td>
      <td>-28.225437</td>
      <td>2.850430e-175</td>
      <td>8.149380e-172</td>
      <td>chr3R</td>
      <td>26869237</td>
      <td>26871995</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>BM-40-SPARC</td>
      <td>¬†</td>
    </tr>
  </tbody>
</table>

<p>‚Äã     b. To check out the last three rows, we would use the command, <code class="language-plaintext highlighter-rouge">data.tail(n=3)</code>, analogous to <code class="language-plaintext highlighter-rouge">head()</code> used above. However, here we want to look 	at the last three columns so we need to change our view and then use <code class="language-plaintext highlighter-rouge">tail()</code>. To do so, we create a new DataFrame in which rows and columns are 	switched:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>data_flipped = data.T
</code></pre></div></div>

<p>‚Äã	We can then view the last three columns of the data by viewing the last three rows of data_flipped:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>data_flipped.tail(n=3)
</code></pre></div></div>

<table>
  <thead>
    <tr>
      <th>GeneID</th>
      <th>FBgn0039155</th>
      <th>FBgn0003360</th>
      <th>FBgn0026562</th>
      <th>FBgn0025111</th>
      <th>FBgn0029167</th>
      <th>FBgn0039827</th>
      <th>FBgn0035085</th>
      <th>FBgn0034736</th>
      <th>FBgn0264475</th>
      <th>FBgn0000071</th>
      <th>‚Ä¶</th>
      <th>FBgn0264343</th>
      <th>FBgn0038237</th>
      <th>FBgn0020376</th>
      <th>FBgn0028939</th>
      <th>FBgn0036560</th>
      <th>FBgn0035710</th>
      <th>FBgn0035523</th>
      <th>FBgn0038261</th>
      <th>FBgn0039178</th>
      <th>FBgn0034636</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Strand</td>
      <td>+</td>
      <td>-</td>
      <td>-</td>
      <td>-</td>
      <td>+</td>
      <td>+</td>
      <td>+</td>
      <td>+</td>
      <td>+</td>
      <td>+</td>
      <td>‚Ä¶</td>
      <td>+</td>
      <td>-</td>
      <td>+</td>
      <td>+</td>
      <td>+</td>
      <td>-</td>
      <td>+</td>
      <td>+</td>
      <td>+</td>
      <td>-</td>
    </tr>
    <tr>
      <td>Feature</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>linc<span class="notranslate">RNA</span></td>
      <td>protein_coding</td>
      <td>‚Ä¶</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
      <td>protein_coding</td>
    </tr>
    <tr>
      <td>Gene name</td>
      <td>Kal1</td>
      <td>sesB</td>
      <td>BM-40-SPARC</td>
      <td>Ant2</td>
      <td>Hml</td>
      <td>CG1544</td>
      <td>CG3770</td>
      <td>CG6018</td>
      <td>CR43883</td>
      <td>Ama</td>
      <td>‚Ä¶</td>
      <td>CG43799</td>
      <td>Pde6</td>
      <td>Sr-CIII</td>
      <td>NimC2</td>
      <td>CG5895</td>
      <td>SP1173</td>
      <td>CG1311</td>
      <td>CG14856</td>
      <td>CG6356</td>
      <td>CG10440</td>
    </tr>
  </tbody>
</table>

<ul>
  <li>Note about Pandas DataFrames/Series</li>
</ul>

<p>A <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html">DataFrame</a> is a collection of <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html">Series</a>; The DataFrame is the way Pandas represents a table, and Series is the data-structure Pandas use to represent a column.</p>

<p>Pandas is built on top of the Numpy library, which in practice means that most of the methods defined for Numpy Arrays apply to Pandas Series/DataFrames.</p>

<p>What makes Pandas so attractive is the powerful interface to access individual records of the table, proper handling of missing values, and relational-databases operations between DataFrames.</p>

<h2 id="select-data">Select data</h2>
<p>To access a value at the position <code class="language-plaintext highlighter-rouge">[i,j]</code> of a DataFrame, we have two options, depending on what is the meaning of i in use. Remember that a DataFrame provides an index as a way to identify the rows of the table; a row, then, has a position inside the table as well as a label, which uniquely identifies its entry in the DataFrame.</p>

<p>You can use <code class="language-plaintext highlighter-rouge">DataFrame.iloc[..., ...]</code> to select values by their (entry) position and basically specify location by numerical index analogously to 2D version of character selection in strings.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
</code></pre></div></div>
<pre><code class="language-output">1086.974295204891087
</code></pre>
<p>You can also use <code class="language-plaintext highlighter-rouge">DataFrame.loc[..., ...]</code> to select values by their (entry) label and basically specify location by row name analogously to 2D version of dictionary keys.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="s">"FBgn0039155"</span><span class="p">,</span> <span class="s">"Base mean"</span><span class="p">])</span>
</code></pre></div></div>
<pre><code class="language-output">1086.97429520489
</code></pre>
<p>You can use Python‚Äôs usual slicing notation, to select all or a subset of rows and/or columns. For example, the following code selects all the columns of the row <code class="language-plaintext highlighter-rouge">"FBgn0039155"</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="s">"FBgn0039155"</span><span class="p">,</span> <span class="p">:])</span>
</code></pre></div></div>
<pre><code class="language-output">Base mean        1086.974295
log2(FC)            -4.14845
StdErr              0.134949
Wald-Stats        -30.740913
P-value                  0.0
P-adj                    0.0
Chromosome             chr3R
Start               24141394
End                 24147490
Strand                     +
Feature       protein_coding
Gene name               Kal1
Name: FBgn0039155, dtype: object 
</code></pre>
<p>Which would get the same result as printing <code class="language-plaintext highlighter-rouge">data.loc["FBgn0039155"]</code> (without a second index).</p>

<p>You can select multiple columns or rows using <code class="language-plaintext highlighter-rouge">DataFrame.loc</code> and a named slice or <code class="language-plaintext highlighter-rouge">Dataframe.iloc</code> and the numbers corresponding to the rows and columns.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="s">'FBgn0003360'</span><span class="p">:</span><span class="s">'FBgn0029167'</span><span class="p">,</span> <span class="s">'Base mean'</span><span class="p">:</span><span class="s">'Wald-Stats'</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span> <span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
</code></pre></div></div>
<pre><code class="language-output">                Base mean  log2(FC)    StdErr  Wald-Stats
GeneID                                                   
FBgn0003360   6409.577128 -2.999777  0.104345  -28.748637
FBgn0026562  65114.840564 -2.380164  0.084327  -28.225437
FBgn0025111   2192.322369  2.699939  0.097945   27.565978
FBgn0029167   5430.067277 -2.105062  0.092547  -22.745964

                Base mean  log2(FC)    StdErr
GeneID                                       
FBgn0003360   6409.577128 -2.999777  0.104345
FBgn0026562  65114.840564 -2.380164  0.084327
FBgn0025111   2192.322369  2.699939  0.097945
</code></pre>

<ul>
  <li>Note the difference between the 2 outputs.</li>
</ul>

<p>When choosing or transitioning between <code class="language-plaintext highlighter-rouge">loc</code> and <code class="language-plaintext highlighter-rouge">iloc</code>, you should keep in mind that the two methods use slightly different indexing schemes.</p>

<p><code class="language-plaintext highlighter-rouge">iloc</code> uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So <code class="language-plaintext highlighter-rouge">0:10</code> will select entries <code class="language-plaintext highlighter-rouge">0,...,9</code>. <code class="language-plaintext highlighter-rouge">loc</code>, meanwhile, indexes inclusively. So <code class="language-plaintext highlighter-rouge">0:10</code> will select entries <code class="language-plaintext highlighter-rouge">0,...,10</code>.</p>

<p>This is particularly confusing when the DataFrame index is a simple numerical list, e.g. <code class="language-plaintext highlighter-rouge">0,...,1000</code>. In this case <code class="language-plaintext highlighter-rouge">df.iloc[0:1000]</code> will return 1000 entries, while <code class="language-plaintext highlighter-rouge">df.loc[0:1000]</code> return 1001 of them! To get 1000 elements using <code class="language-plaintext highlighter-rouge">loc</code>, you will need to go one lower and ask for <code class="language-plaintext highlighter-rouge">df.loc[0:999]</code>.</p>

<p>The result of slicing is a new dataframe and can be used in further operations. All the statistical operators that work on entire dataframes work the same way on slices. E.g., calculate max of a slice.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="s">'FBgn0003360'</span><span class="p">:</span><span class="s">'FBgn0029167'</span><span class="p">,</span> <span class="s">'Base mean'</span><span class="p">].</span><span class="nb">max</span><span class="p">())</span>
</code></pre></div></div>
<pre><code class="language-output">65114.8405637953
</code></pre>

<h2 id="use-conditionals-to-select-data">Use conditionals to select data</h2>

<p>You can use conditionals to select data. A comparison is applied element by element and returns a similarly-shaped dataframe of <code class="language-plaintext highlighter-rouge">True</code> and <code class="language-plaintext highlighter-rouge">False</code>. The last one can be used as a mask to subset the original dataframe. The following example creates a new dataframe consisting only of the columns ‚ÄòP-adj‚Äô and ‚ÄòGene name‚Äô, then keeps the rows that comply with the <span class="notranslate">expression</span> <code class="language-plaintext highlighter-rouge">'P-adj' &lt; 0.000005</code></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">subset</span> <span class="o">=</span> <span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s">'P-adj'</span><span class="p">,</span> <span class="s">'Gene name'</span><span class="p">]]</span>
<span class="k">print</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">                     P-adj    Gene name
GeneID                                 
FBgn0039155  1.387207e-203         Kal1
FBgn0003360  4.039734e-178         sesB
FBgn0026562  8.149380e-172  BM-40-SPARC
FBgn0025111  6.105789e-164         Ant2
FBgn0029167  2.698811e-111          Hml
...                    ...          ...
FBgn0035710   1.436480e-04       SP1173
FBgn0035523   1.523189e-04       CG1311
FBgn0038261   1.525142e-04      CG14856
FBgn0039178   1.629061e-04       CG6356
FBgn0034636   1.951192e-04      CG10440

[130 rows x 2 columns]
</code></pre>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mask</span> <span class="o">=</span> <span class="n">subset</span><span class="p">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s">'P-adj'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.000005</span>
<span class="n">new_data</span> <span class="o">=</span> <span class="n">subset</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="n">new_data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">
                     P-adj    Gene name
GeneID                                 
FBgn0039155  1.387207e-203         Kal1
FBgn0003360  4.039734e-178         sesB
FBgn0026562  8.149380e-172  BM-40-SPARC
FBgn0025111  6.105789e-164         Ant2
FBgn0029167  2.698811e-111          Hml
...                    ...          ...
FBgn0039593   1.544012e-07       CG9989
FBgn0265512   3.413938e-07          mlt
FBgn0030326   3.625857e-07       CG2444
FBgn0039485   7.174821e-07      CG17189
FBgn0025836   1.071884e-06     RhoGAP1A

[113 rows x 2 columns]
</code></pre>

<p>If we have not had specified the column, that the <span class="notranslate">expression</span> should be applied to, then it would have been applied to the entire dataframe. But the dataframe contains different type of data. In that case, an error would occur.</p>

<p>Consider the following example of a dataframe consisting only of numerical data. The <span class="notranslate">expression</span> and the mask would be normally applied to the data and the mask would return <code class="language-plaintext highlighter-rouge">NaN</code> for the data that don‚Äôt comply with the <span class="notranslate">expression</span>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">subset</span> <span class="o">=</span> <span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s">'StdErr'</span><span class="p">,</span>	<span class="s">'Wald-Stats'</span><span class="p">,</span> <span class="s">'P-value'</span><span class="p">,</span> <span class="s">'P-adj'</span><span class="p">]]</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">subset</span> <span class="o">&lt;</span> <span class="mf">0.05</span>
<span class="n">new_data</span> <span class="o">=</span> <span class="n">subset</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="n">new_data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">             StdErr  Wald-Stats        P-value          P-adj
GeneID                                                       
FBgn0039155     NaN  -30.740913  1.617357e-207  1.387207e-203
FBgn0003360     NaN  -28.748637  9.419922e-182  4.039734e-178
FBgn0026562     NaN  -28.225437  2.850430e-175  8.149380e-172
FBgn0025111     NaN         NaN  2.847517e-167  6.105789e-164
FBgn0029167     NaN  -22.745964  1.573284e-114  2.698811e-111
...             ...         ...            ...            ...
FBgn0035710     NaN         NaN   6.682473e-06   1.436480e-04
FBgn0035523     NaN         NaN   7.103599e-06   1.523189e-04
FBgn0038261     NaN         NaN   7.130487e-06   1.525142e-04
FBgn0039178     NaN         NaN   7.654328e-06   1.629061e-04
FBgn0034636     NaN   -4.430118   9.418135e-06   1.951192e-04

[130 rows x 4 columns]
</code></pre>
<p>This is very useful because NaNs are ignored by operations like max, min, average, etc.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">new_data</span><span class="p">.</span><span class="n">describe</span><span class="p">())</span>
</code></pre></div></div>
<pre><code class="language-output">       StdErr  Wald-Stats        P-value          P-adj
count     0.0   70.000000   1.300000e+02   1.300000e+02
mean      NaN  -11.035514   4.200552e-07   9.321175e-06
std       NaN    5.695077   1.531253e-06   3.278527e-05
min       NaN  -30.740913  1.617357e-207  1.387207e-203
25%       NaN  -13.059226   4.952408e-31   1.275950e-28
50%       NaN   -9.971703   2.109219e-19   2.563262e-17
75%       NaN   -6.963977   6.731953e-12   3.706020e-10
max       NaN   -4.430118   9.418135e-06   1.951192e-04
</code></pre>

<h3 id="exercises-1">Exercises</h3>
<ol>
  <li>
    <p>Explain what each line in the following short program does: what is in first, second, etc.?</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">first</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://zenodo.org/record/3477564/files/annotatedDEgenes.tabular"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="s">'GeneID'</span><span class="p">)</span>
 <span class="n">second</span> <span class="o">=</span> <span class="n">first</span><span class="p">[</span><span class="n">first</span><span class="p">[</span><span class="s">'log2(FC)'</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">]</span>
 <span class="n">third</span> <span class="o">=</span> <span class="n">second</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'FBgn0025111'</span><span class="p">)</span>
 <span class="n">fourth</span> <span class="o">=</span> <span class="n">third</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'StdErr'</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
 <span class="n">fourth</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">'result.csv'</span><span class="p">)</span>
</code></pre></div>    </div>
  </li>
  <li>
    <p>Explain in simple terms what <code class="language-plaintext highlighter-rouge">idxmin</code> and <code class="language-plaintext highlighter-rouge">idxmax</code> do in the short program below. When would you use these methods?</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://zenodo.org/record/3477564/files/annotatedDEgenes.tabular"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="s">'GeneID'</span><span class="p">)</span>
    
<span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">'Base mean'</span><span class="p">].</span><span class="n">idxmin</span><span class="p">())</span>
<span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">'Base mean'</span><span class="p">].</span><span class="n">idxmax</span><span class="p">())</span>
    
</code></pre></div>    </div>
  </li>
  <li>
    <p>Assume Pandas has been imported and the previous dataset has been loaded. Write an <span class="notranslate">expression</span> to select each of the following:</p>

    <p>a. P-value of each gene
b. all the information of gene <code class="language-plaintext highlighter-rouge">FBgn0039178</code>
c. the information of all genes that belong to chromosome <code class="language-plaintext highlighter-rouge">chr3R</code></p>
  </li>
</ol>

<h3 id="solutions-1">Solutions</h3>
<ol>
  <li>
    <p>Let‚Äôs go through this piece of code line by line.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">first</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://zenodo.org/record/3477564/files/annotatedDEgenes.tabular"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="s">'GeneID'</span><span class="p">)</span>
</code></pre></div>    </div>
    <p>This line loads the data into a dataframe called first. The <code class="language-plaintext highlighter-rouge">index_col='GeneID'</code> parameter selects which column to use as the row labels in the dataframe.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">second</span> <span class="o">=</span> <span class="n">first</span><span class="p">[</span><span class="n">first</span><span class="p">[</span><span class="s">'log2(FC)'</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">]</span>
</code></pre></div>    </div>
    <p>This line makes a selection: only those rows of first for which the ‚Äòlog2(FC)‚Äô column contains a positive value are extracted. Notice how the Boolean <span class="notranslate">expression</span> inside the brackets is used to select only those rows where the <span class="notranslate">expression</span> is true.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">third</span> <span class="o">=</span> <span class="n">second</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'FBgn0025111'</span><span class="p">)</span>
</code></pre></div>    </div>
    <p>As the syntax suggests, this line drops the row from second where the label is ‚ÄòFBgn0025111‚Äô. The resulting dataframe third has one row less than the original dataframe second.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">fourth</span> <span class="o">=</span> <span class="n">third</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'StdErr'</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</code></pre></div>    </div>
    <p>Again we apply the drop function, but in this case we are dropping not a row but a whole column. To accomplish this, we need to specify also the axis parameter.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">fourth</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">'result.csv'</span><span class="p">)</span>
</code></pre></div>    </div>
    <p>The final step is to write the data that we have been working on to a csv file. Pandas makes this easy with the <code class="language-plaintext highlighter-rouge">to_csv()</code> function. The only required argument to the function is the filename. Note that the file will be written in the directory from which you started the Jupyter or Python session.</p>
  </li>
  <li>
    <p><code class="language-plaintext highlighter-rouge">idxmin</code> will return the index value corresponding to the minimum; idxmax will do the same for the maximum value.</p>

    <p>You can use these functions whenever you want to get the row index of the minimum/maximum value and not the actual minimum/maximum value.</p>

    <pre><code class="language-output"> FBgn0063667
 FBgn0026562
</code></pre>
  </li>
  <li>
    <p>a.</p>

    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">data</span><span class="p">[</span><span class="s">'P-value'</span><span class="p">]</span>
</code></pre></div>    </div>
    <p>b.</p>
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   <span class="n">data</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="s">'FBgn0039178'</span><span class="p">,</span> <span class="p">:]</span>
</code></pre></div>    </div>
    <p>c.</p>
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s">'Chromosome'</span><span class="p">]</span> <span class="o">==</span> <span class="s">'chr3R'</span><span class="p">]</span>
</code></pre></div>    </div>
  </li>
</ol>

<h2 id="group-by-and-analyze-the-data">Group-by and analyze the data</h2>

<p>Many data analysis tasks can be approached using the ‚Äúsplit-apply-combine‚Äù paradigm: split the data into groups, apply some analysis to each group, and then combine the results.</p>

<p>Pandas makes this very easy through the use of the <code class="language-plaintext highlighter-rouge">groupby()</code> method, which splits the data into groups. When the data is grouped in this way, the aggregate method <code class="language-plaintext highlighter-rouge">agg()</code> can be used to apply an aggregating or summary function to each group.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">summarised_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="s">'Chromosome'</span><span class="p">).</span><span class="n">agg</span><span class="p">({</span><span class="s">'Base mean'</span><span class="p">:</span><span class="s">'first'</span><span class="p">,</span> 
                             <span class="s">'log2(FC)'</span><span class="p">:</span> <span class="s">'max'</span><span class="p">})</span>
<span class="k">print</span><span class="p">(</span><span class="n">summarised_data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">              Base mean  log2(FC)
Chromosome                       
chr2L       1269.362573  2.145440
chr2R        928.263812  2.407986
chr3L       5430.067277  2.428607
chr3R       1086.974295  2.360017
chrX        6409.577128  2.699939
</code></pre>

<p>There are a couple of things that should be noted. The <code class="language-plaintext highlighter-rouge">agg()</code> method accepts a dictionary as input that specifies the function to be applied to each column. The output is a new dataframe, that each row corresponds to one group. The output dataframe uses the grouping column as index. We could change the last one by simply using the <code class="language-plaintext highlighter-rouge">reset_index()</code> method.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">summarised_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="s">'Chromosome'</span><span class="p">).</span><span class="n">agg</span><span class="p">({</span><span class="s">'Base mean'</span><span class="p">:</span><span class="s">'first'</span><span class="p">,</span> 
                                                  <span class="s">'log2(FC)'</span><span class="p">:</span> <span class="s">'max'</span><span class="p">}).</span><span class="n">reset_index</span><span class="p">()</span> 
<span class="k">print</span><span class="p">(</span><span class="n">summarised_data</span><span class="p">)</span>
</code></pre></div></div>
<pre><code class="language-output">  Chromosome    Base mean  log2(FC)
0      chr2L  1269.362573  2.145440
1      chr2R   928.263812  2.407986
2      chr3L  5430.067277  2.428607
3      chr3R  1086.974295  2.360017
4       chrX  6409.577128  2.699939
</code></pre>
<h3 id="exercises-2">Exercises</h3>

<ol>
  <li>Import the tabular data from the ‚Äúhttps://zenodo.org/record/3477564/files/annotatedDEgenes.tabular‚Äù link. What are the longest genes in each chromosome?</li>
  <li>Using the same dataset,  try to find how many genes are found on each strand of each chromosome.</li>
</ol>

<h3 id="solutions-2">Solutions</h3>
<ol>
  <li>
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://zenodo.org/record/3477564/files/annotatedDEgenes.tabular"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="s">'GeneID'</span><span class="p">)</span>
    
<span class="n">data</span><span class="p">[</span><span class="s">'Gene Length'</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s">'End'</span><span class="p">]</span> <span class="o">-</span> <span class="n">data</span><span class="p">[</span><span class="s">'Start'</span><span class="p">]</span>
<span class="n">data</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="s">'Chromosome'</span><span class="p">).</span><span class="n">agg</span><span class="p">(</span><span class="n">max_length</span> <span class="o">=</span> <span class="p">(</span><span class="s">'Gene Length'</span><span class="p">,</span> <span class="s">'max'</span><span class="p">))</span>
</code></pre></div>    </div>
  </li>
  <li>You can group the data according to more than one column.
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="n">data</span><span class="p">.</span><span class="n">groupby</span><span class="p">([</span><span class="s">'Chromosome'</span><span class="p">,</span> <span class="s">'Strand'</span><span class="p">]).</span><span class="n">size</span><span class="p">()</span>
</code></pre></div>    </div>
  </li>
</ol>

<h1 id="conclusion">Conclusion</h1>

<p>This tutorial aims to serve as an introduction to data analysis using the Python programming language.</p>

<h2 id="key-points">Key points</h2>

<ul>
  <li>
    <p>Python has many libraries offering a variety of capabilities, which makes it popular for beginners, as well as, more experienced users</p>
  </li>
  <li>
    <p>You can use scientific libraries like Numpy and Pandas to perform data analysis.</p>
  </li>
</ul>

<h1 id="references">References</h1>

<ol>
  <li>Wendi Bacon, Mehmet Tekman, 2021, Trajectory Analysis using Python (Jupyter Notebook) in <span class="notranslate">Galaxy</span> (<span class="notranslate">Galaxy</span> Training Materials). https://training.<span class="notranslate">galaxy</span>project.org/training-material/topics/transcriptomics/tutorials/sc<span class="notranslate">rna</span>-JUPYTER-trajectories/tutorial.html Online; accessed Wed Sep 01 2021</li>
  <li>Batut et al., 2018, Community-Driven Data Analysis Training for Biology Cell Systems [10.1016/j.cels.2018.05.012</li>
  <li>Azalee Bostroem, Trevor Bekolay, and Valentina Staneva (eds): ‚ÄúSoftware Carpentry: Programming with Python.‚Äù  Version 2016.06, June
2016, https://github.com/swcarpentry/python-novice-inflammation, 10.5281/zenodo.57492.</li>
  <li>Allen Lee, Nathan Moore, Sourav Singh, Olav Vahtras (eds). Software Carpentry: Plotting and Programming in Python. http://github.com/swcarpentry/python-novice-plotting, 2018.</li>
  <li>Python 3.9.7 Documentation, https://docs.python.org/3/tutorial/datastructures.html Online; accessed Wed Sep 01 2021</li>
  <li>Karlijn Willems, Pandas Tutorial: DataFrames in Python. https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python Online; accessed Wed Sep 01 2021</li>
  <li>B√©r√©nice Batut, Mallory Freeberg, Mo Heydarian, Anika Erxleben, Pavankumar Videm, Clemens Blank, Maria Doyle, Nicola Soranzo, Peter van Heusden, 2021 Reference-based <span class="notranslate">RNA</span>-Seq data analysis (<span class="notranslate">Galaxy</span> Training Materials). https://training.<span class="notranslate">galaxy</span>project.org/training-material/topics/transcriptomics/tutorials/ref-based/tutorial.html Online; accessed Wed Sep 01 2021</li>
  <li>B√©r√©nice Batut, Fotis E. Psomopoulos, Toby Hodges, 2021 Advanced R in <span class="notranslate">Galaxy</span> (<span class="notranslate">Galaxy</span> Training Materials). https://training.<span class="notranslate">galaxy</span>project.org/training-material/topics/introduction/tutorials/r-advanced/tutorial.html Online; accessed Wed Sep 01 2021</li>
</ol>



                
                <blockquote class="key_points">
                    <h3><i class="fas fa-key" aria-hidden="true"></i> Key points</h3>
                    <ul>
                        
                    </ul>
                </blockquote>
                

                <h1>Frequently Asked Questions</h1>
                Have questions about this tutorial? Check out the  <a href="/training-material/topics/introduction/faqs/">FAQ page for the Introduction to Galaxy Analyses topic</a> to see if your question is listed there.
                If not, please ask your question on the <a href="https://gitter.im/Galaxy-Training-Network/Lobby">GTN Gitter Channel</a> or the
                <a href="https://help.galaxyproject.org">Galaxy Help Forum</a>

                

                

                

                <h1>Feedback</h1>
                <p class="text-muted">Did you use this material as an instructor? Feel free to give us feedback on <a href="https://github.com/galaxyproject/training-material/issues/1452" target="_blank">how it went</a>.</p>

                <div id="feedback-button">
                    <img src="/training-material/shared/images/feedback.png" title="Click to activate" alt="Click here to load Google feedback frame" />
                </div>
                <div id="feedback-form">
                </div>
                <script type="text/javascript">
                    (function (window, document) {
                        function onDocumentReady(fn) {
                            if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading") {
                                fn();
                            } else {
                                document.addEventListener('DOMContentLoaded', fn);
                            }
                        }

                        onDocumentReady(function () {
                            $("#feedback-button").click(function(evt){
                                var e = $(evt.target)
                                e.hide();

                                $("#feedback-form").html(`
                                    <iframe id="feedback-google" class="google-form" src="https://docs.google.com/forms/d/e/1FAIpQLSd4VZptFTQ03kHkMz0JyW9b6_S8geU5KjNE_tLM0dixT3ZQmA/viewform?embedded=true&entry.1235803833=Advanced Python (Introduction to Galaxy Analyses)">Loading...</iframe>
                                `)
                            })
                        });
                    })(window, document);
                </script>



                <h1>Citing this Tutorial</h1>
                <p>
                    <ol>
                        <li id="citation-text">Maria Christina Maniou, Fotis E. Psomopoulos, 2021 <b>Advanced Python (Galaxy Training Materials)</b>. <a href="https://training.galaxyproject.org/training-material/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html">https://training.galaxyproject.org/training-material/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html</a> Online; accessed TODAY
                        </li>
                        <li>
                        Batut et al., 2018 <b>Community-Driven Data Analysis Training for Biology</b> Cell Systems <a href="https://doi.org/10.1016%2Fj.cels.2018.05.012">10.1016/j.cels.2018.05.012</a>
                        </li>
                    </ol>
                </p>


                <blockquote class="details">
                  <h3><i class="fa fa-info-circle" aria-hidden="true"></i><span class="visually-hidden">details</span> BibTeX</h3>
                  <p style="display: none;">

                <div class="highlighter-rouge"><div class="highlight"><pre class="highlight">
<code id="citation-code">@misc{introduction-python-advanced-np-pd,
author = "Maria Christina Maniou and Fotis E. Psomopoulos",
title = "Advanced Python (Galaxy Training Materials)",
year = "2021",
month = "10",
day = "26"
url = "\url{https://training.galaxyproject.org/training-material/topics/introduction/tutorials/python-advanced-np-pd/tutorial.html}",
note = "[Online; accessed TODAY]"
}
@article{Batut_2018,
    doi = {10.1016/j.cels.2018.05.012},
    url = {https://doi.org/10.1016%2Fj.cels.2018.05.012},
    year = 2018,
    month = {jun},
    publisher = {Elsevier {BV}},
    volume = {6},
    number = {6},
    pages = {752--758.e1},
    author = {B{\'{e}}r{\'{e}}nice Batut and Saskia Hiltemann and Andrea Bagnacani and Dannon Baker and Vivek Bhardwaj and Clemens Blank and Anthony Bretaudeau and Loraine Brillet-Gu{\'{e}}guen and Martin {\v{C}}ech and John Chilton and Dave Clements and Olivia Doppelt-Azeroual and Anika Erxleben and Mallory Ann Freeberg and Simon Gladman and Youri Hoogstrate and Hans-Rudolf Hotz and Torsten Houwaart and Pratik Jagtap and Delphine Larivi{\`{e}}re and Gildas Le Corguill{\'{e}} and Thomas Manke and Fabien Mareuil and Fidel Ram{\'{\i}}rez and Devon Ryan and Florian Christoph Sigloch and Nicola Soranzo and Joachim Wolff and Pavankumar Videm and Markus Wolfien and Aisanjiang Wubuli and Dilmurat Yusuf and James Taylor and Rolf Backofen and Anton Nekrutenko and Bj√∂rn Gr√ºning},
    title = {Community-Driven Data Analysis Training for Biology},
    journal = {Cell Systems}
}</code>
                </pre></div></div>
                </p>
                </blockquote>


<script type="text/javascript">
// update the date on load, or leave fallback of 'today'
d = new Date();
document.getElementById("citation-code").innerHTML = document.getElementById("citation-code").innerHTML.replace("TODAY", d.toDateString());
document.getElementById("citation-text").innerHTML = document.getElementById("citation-text").innerHTML.replace("TODAY", d.toDateString());
</script>

                <h3><i class="far fa-thumbs-up" aria-hidden="true"></i> Congratulations on successfully completing this tutorial!</h3>

                

                
                <blockquote class="agenda follow-up">
                    <strong class="follow-up"><i class="fas fa-graduation-cap" aria-hidden="true"></i> Do you want to extend your knowledge? Follow one of our recommended follow-up trainings:</strong>
                    <ul>
                        
                    </ul>
                </blockquote>
                

            </div>
        </div>
    </div>
</section>
<br/>
<br/>
<br/>

        </div>
        
    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap-toc.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/theme.js"></script>

    <script type="text/javascript" src="/training-material/assets/js/clipboard.min.js"></script>
    <script type="text/javascript">
    var snippets=document.querySelectorAll('div.highlight');
    [].forEach.call(snippets,function(snippet){
        snippet.firstChild.insertAdjacentHTML('beforebegin','<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets=new ClipboardJS('[data-clipboard-snippet]',{
        target:function(trigger){return trigger.nextElementSibling;
    }});
    </script>
    

    <script type="text/javascript">
        if(window.location.hostname === "galaxyproject.github.io") {
            // Redirect
            var redirect = "https://training.galaxyproject.org" + window.location.pathname + window.location.search;
            $('div.container.main-content').prepend("<div class='alert alert-warning'><strong>Note: </strong>This content has a new home at <a href=\"" + redirect + "\">" + redirect + "</a>, which you will be redirected to in 5 seconds.</div>");

            window.setTimeout(function(){
                window.location.href = redirect;
            }, 5000)

        }
    </script>
</html>
